{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Runs through and finds the most conserved (rbfox2) motifs\n",
    "- comments out some of these blocks since they're a part of scoring/file creation which takes awhile. Once the files are made, no need to run through again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import pybedtools\n",
    "import pyBigWig\n",
    "import os\n",
    "from scipy import stats\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/01_SAILOR_bulk_rnaseq/outputs/motif_coords/'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/01_SAILOR_bulk_rnaseq/outputs/motif_coords/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "phastcons_file = '/projects/ps-yeolab3/bay001/annotations/conservation/hg19.phastCons46way.placental.wigFix.bw'\n",
    "motifs_file = os.path.join(input_dir, 'GCATG_RBFOX_MOTIF_POSITIONS_ALL.annotated') # these were annotated in a strand-specific manner, so positions on the opposite strand of a gene will mostly be labeled intergenic (unless there is an overlapping antisense thing going on)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use a modified function to get conservation scores from bigwig file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Density:\n",
    "    def values(self, chrom, start, end, strand):\n",
    "        return 0\n",
    "    \n",
    "class Phastcon(Density):\n",
    "\n",
    "    def __init__(self, phastcon, name=None):\n",
    "        try:\n",
    "            self.phastcon = pyBigWig.open(phastcon)\n",
    "            self.name = name if name is not None else os.path.basename(phastcon)\n",
    "        except Exception as e:\n",
    "            print(\"couldn't open the bigwig file!\")\n",
    "            print(e)\n",
    "\n",
    "    def values(self, chrom, start, end, strand):\n",
    "        \"\"\"\n",
    "        Parameters\n",
    "        ----------\n",
    "        chrom : basestring\n",
    "            (eg. chr1)\n",
    "        start : int\n",
    "            0-based start (first position in chromosome is 0)\n",
    "        end : int\n",
    "            1-based end (last position is not included)\n",
    "        strand : str\n",
    "            either '+' or '-'\n",
    "        Returns\n",
    "        -------\n",
    "        densites : list\n",
    "            values corresponding to density over specified positions.\n",
    "        \"\"\"\n",
    "\n",
    "        try:\n",
    "            if strand == \"+\":\n",
    "                return self.phastcon.values(chrom, start, end)\n",
    "            elif strand == \"-\":\n",
    "                return list(reversed(self.phastcon.values(chrom, start, end)))\n",
    "            else:\n",
    "                print(\"Strand neither + or -\")\n",
    "                return 1\n",
    "        except RuntimeError:\n",
    "            # usually occurs when no chromosome exists in the bigwig file\n",
    "            return [np.NaN] * abs(start - end)\n",
    "        \n",
    "    def avg_cons(self, chrom, start, end, strand):\n",
    "        \n",
    "        values = pd.Series(self.values(chrom, start, end, strand))\n",
    "        values.fillna(0, inplace=True)\n",
    "        return values.mean()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd = Phastcon(phastcons_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.9309999942779541,\n",
       " 0.9340000152587891,\n",
       " 0.9350000023841858,\n",
       " 0.9350000023841858,\n",
       " 0.9330000281333923]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cd.values('chrY',28438600,28438605,'-') # 'chr10:180750-180776'chrY \t28438600 \t28438605"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motifs file contains every position where I found either GCATG or its reverse complement (CATGC). \n",
    "- Let's try to use annotator to select correctly-stranded regions. If we cannot identify a position on its specified strand, can we assume that it's not really a true motif?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "annotated_names = [\n",
    "    'chrom','start','end','name','score','strand',\n",
    "    'geneid','genename','region','annotation'\n",
    "]\n",
    "motifs = pd.read_csv(motifs_file, sep='\\t', names=annotated_names)\n",
    "motifs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's just test to make sure we can definitively say the above assumption is true."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "motifs = motifs[['chrom','start','end','geneid','score','strand']]\n",
    "\n",
    "genic_motifs = motifs[motifs['geneid']!='intergenic']\n",
    "intergenic_motifs = motifs[motifs['geneid']=='intergenic']\n",
    "\n",
    "print(\"Found {} genic motifs and {} intergenic motifs\".format(genic_motifs.shape[0], intergenic_motifs.shape[0]))\n",
    "\n",
    "# Found 3395630 genic motifs and 8023621 intergenic motifs"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genic_motifs_bt = pybedtools.BedTool.from_dataframe(genic_motifs)\n",
    "intergenic_motifs_bt = pybedtools.BedTool.from_dataframe(intergenic_motifs)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def stringify(interval):\n",
    "    return \"{}:{}-{}:{}\".format(\n",
    "        interval.chrom,\n",
    "        interval.start,\n",
    "        interval.end,\n",
    "        interval.strand\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genic_phastcon_scores = defaultdict()\n",
    "genic_phastcon_total = 0\n",
    "intergenic_phastcon_scores = defaultdict()\n",
    "intergenic_phastcon_total = 0\n",
    "\n",
    "progress = tnrange(genic_motifs.shape[0])\n",
    "for interval in genic_motifs_bt:\n",
    "    avg_cons = cd.avg_cons(interval.chrom, interval.start, interval.end, interval.strand)\n",
    "    genic_phastcon_scores[stringify(interval)] = avg_cons\n",
    "    genic_phastcon_total += avg_cons\n",
    "    progress.update(1)\n",
    "    \n",
    "progress = tnrange(intergenic_motifs.shape[0])\n",
    "for interval in intergenic_motifs_bt:\n",
    "    avg_cons = cd.avg_cons(interval.chrom, interval.start, interval.end, interval.strand)\n",
    "    intergenic_phastcon_scores[stringify(interval)] = avg_cons\n",
    "    intergenic_phastcon_total += avg_cons\n",
    "    progress.update(1)\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "print(\n",
    "    \"Average genic phastcon score: {}, average intergenic phastcon score: {}\".format(\n",
    "        genic_phastcon_total/genic_motifs.shape[0],\n",
    "        intergenic_phastcon_total/intergenic_motifs.shape[0]\n",
    "    )\n",
    ")\n",
    "# Average genic phastcon score: 0.07799540462852061, average intergenic phastcon score: 0.06755178614135071"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pd.DataFrame(genic_phastcon_scores, index=['scores']).T.to_csv(\n",
    "    os.path.join(output_dir, 'GCATG_RBFOX_MOTIF_POSITIONS_ALL_GENIC.phastCons46way.txt'),\n",
    "    sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pd.DataFrame(intergenic_phastcon_scores, index=['scores']).T.to_csv(\n",
    "    os.path.join(output_dir, 'GCATG_RBFOX_MOTIF_POSITIONS_ALL_INTERGENIC.phastCons46way.txt'),\n",
    "    sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Let's select motifs found on genic (annotated) regions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10:91096-91101:+</th>\n",
       "      <td>0.1272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:91842-91847:+</th>\n",
       "      <td>0.0316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:92302-92307:+</th>\n",
       "      <td>0.0624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:92729-92734:+</th>\n",
       "      <td>0.0274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:180760-180765:+</th>\n",
       "      <td>0.0128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       scores\n",
       "chr10:91096-91101:+    0.1272\n",
       "chr10:91842-91847:+    0.0316\n",
       "chr10:92302-92307:+    0.0624\n",
       "chr10:92729-92734:+    0.0274\n",
       "chr10:180760-180765:+  0.0128"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genic_scores = pd.read_csv(os.path.join(output_dir, 'GCATG_RBFOX_MOTIF_POSITIONS_ALL_GENIC.phastCons46way.txt'), sep='\\t', index_col=0)\n",
    "genic_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1697815\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:91096-91101:+</td>\n",
       "      <td>0.1272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:91842-91847:+</td>\n",
       "      <td>0.0316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:92302-92307:+</td>\n",
       "      <td>0.0624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:92729-92734:+</td>\n",
       "      <td>0.0274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:180760-180765:+</td>\n",
       "      <td>0.0128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   index  scores\n",
       "0    chr10:91096-91101:+  0.1272\n",
       "1    chr10:91842-91847:+  0.0316\n",
       "2    chr10:92302-92307:+  0.0624\n",
       "3    chr10:92729-92734:+  0.0274\n",
       "4  chr10:180760-180765:+  0.0128"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZMAAAEICAYAAACavRnhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAFjZJREFUeJzt3X+w3XV95/HnSyKIVAShZjChBkt0i9CdYgrU7rSpdDGga/xDdnGpRJc2Oy5aLcxWdHeGjqwzdHcpK47aZoUFOlREtrNkFKUMetdtRxCorQiUJSIlkShCgBqpP+K+94/ziR6v5yY353NvDsf7fMycud/v+/v5fj+fz8klr3x/nEOqCkmSejxr0gOQJE0/w0SS1M0wkSR1M0wkSd0ME0lSN8NEktTNMJEkdTNMJEndDBPpGSgD/vepqeEvq9QhybuSfC3Jt5Lcn+TUJAckeU+Sr7T6XUmObu1fmeSOJE+1n68cOtZMkvcl+SvgaeAlSZ6f5Iok21s//ynJAa39sUn+dzvWY0k+Npl3QYJlkx6ANK2SvAx4G/DLVfVIklXAAcD5wBuBM4D/C/wi8HSSFwCfBH4X+ChwJvDJJMdW1ePtsG8CTgfuBwJ8HPgGcCxwCPAJYCvwJ8DFwF8AvwEcCKxZ3BlLc/PMRBrfD4CDgOOSPLuqHqqqrwC/DfzHqrq/Bv62hcVrgAeq6k+raldVfRT4O+BfDB3zqqq6p6p2AS9gECzvrKpvV9WjwGXAWa3t94EXAy+qqu9U1V/ul1lLIxgm0piqagvwTuAPgEeTXJfkRcDRwFdG7PIi4O9n1f4eWDG0vnVo+cXAs4HtSZ5M8iSDM5IXtu2/z+Ds5QtJ7knybzqnJI3NMJE6VNWfVdU/Y/AXfwF/yCAQfn5E80dau2E/B3xt+JBDy1uB7wJHVtVh7XVoVb289f31qvqdqnoR8G+BDyU5dkEmJu0jw0QaU5KXJXlVkoOA7wD/yODS10eAi5Osbk9l/WKSI4CbgJcm+ddJliX5V8BxDO6D/ISq2s7gnsilSQ5N8qwkP5/k11v/ZyZZ2Zo/wSCIfrCYc5bmYphI4zsIuAR4DPg6g8tP7wH+CLieQRD8A3AFcHC7b/Ja4ALgcQaXqV5bVY/toY9zGNxcv5dBYNwAHNW2/TJwe5KdwGbgHVX11YWcoDRf8X+OJUnq5ZmJJKmbYSJJ6maYSJK6GSaSpG5L5utUjjzyyFq1atVY+37729/mkEMOWdgBPcM556XBOS8NPXO+6667Hquqn91buyUTJqtWreLOO+8ca9+ZmRnWrl27sAN6hnPOS4NzXhp65pxk9rc2jORlLklSN8NEktTNMJEkdTNMJEndDBNJUjfDRJLUzTCRJHUzTCRJ3QwTSVI3w2Qe7v7aU6y68JOsuvCTkx6KJD0jGSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSp217DJMmVSR5N8uWh2guS3JLkgfbz8FZPksuTbEnypSQnDu2zobV/IMmGoforktzd9rk8ScbtQ5I0GfM5M7kKWDerdiFwa1WtBm5t6wCnA6vbayPwYRgEA3ARcDJwEnDR7nBobTYO7bdunD4kSZOz1zCpqs8BO2aV1wNXt+WrgdcP1a+pgduAw5IcBbwauKWqdlTVE8AtwLq27dCq+nxVFXDNrGPtSx+SpAlZNuZ+y6tqO0BVbU/ywlZfAWwdaret1fZU3zaiPk4f22cPMslGBmcvLF++nJmZmX2b5e6BHAwXnLALYOxjTJudO3cumbnu5pyXBue8OMYNk7lkRK3GqI/Tx08WqzYBmwDWrFlTa9eu3cuhR/vAtTdy6d2Dt+qhs8c7xrSZmZlh3PdrWjnnpcE5L45xn+b6xu5LS+3no62+DTh6qN1K4JG91FeOqI/ThyRpQsYNk83A7ieyNgA3DtXPaU9cnQI81S5V3QycluTwduP9NODmtu1bSU5pT3GdM+tY+9KHJGlC9nqZK8lHgbXAkUm2MXgq6xLg+iTnAg8DZ7bmNwFnAFuAp4G3AFTVjiQXA3e0du+tqt039d/K4Imxg4FPtRf72ockaXL2GiZV9cY5Np06om0B581xnCuBK0fU7wSOH1F/fF/7kCRNhp+AlyR1M0wkSd0ME0lSN8NEktTNMJEkdTNMJEndDBNJUjfDRJLUzTCRJHUzTCRJ3QwTSVI3w0SS1M0wkSR1M0wkSd0ME0lSN8NEktTNMJEkdTNMJEndDBNJUjfDRJLUzTCRJHUzTCRJ3QwTSVI3w0SS1M0wkSR1M0wkSd0ME0lSN8NEktTNMJEkdesKkyS/l+SeJF9O8tEkz0lyTJLbkzyQ5GNJDmxtD2rrW9r2VUPHeXer35/k1UP1da22JcmFQ/WRfUiSJmPsMEmyAvhdYE1VHQ8cAJwF/CFwWVWtBp4Azm27nAs8UVXHApe1diQ5ru33cmAd8KEkByQ5APggcDpwHPDG1pY99CFJmoDey1zLgIOTLAOeC2wHXgXc0LZfDby+La9v67TtpyZJq19XVd+tqq8CW4CT2mtLVT1YVd8DrgPWt33m6kOSNAHLxt2xqr6W5L8CDwP/CPwFcBfwZFXtas22ASva8gpga9t3V5KngCNa/bahQw/vs3VW/eS2z1x9/JgkG4GNAMuXL2dmZmasuS4/GC44YdDduMeYNjt37lwyc93NOS8NznlxjB0mSQ5ncFZxDPAk8HEGl6Rmq927zLFtrvqos6Y9tf/JYtUmYBPAmjVrau3ataOa7dUHrr2RS+8evFUPnT3eMabNzMwM475f08o5Lw3OeXH0XOb6TeCrVfXNqvo+8OfAK4HD2mUvgJXAI215G3A0QNv+fGDHcH3WPnPVH9tDH5KkCegJk4eBU5I8t93HOBW4F/gs8IbWZgNwY1ve3NZp2z9TVdXqZ7WnvY4BVgNfAO4AVrcntw5kcJN+c9tnrj4kSRMwdphU1e0MboL/NXB3O9Ym4F3A+Um2MLi/cUXb5QrgiFY/H7iwHece4HoGQfRp4Lyq+kG7J/I24GbgPuD61pY99CFJmoCx75kAVNVFwEWzyg8yeBJrdtvvAGfOcZz3Ae8bUb8JuGlEfWQfkqTJ8BPwkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkrp1hUmSw5LckOTvktyX5FeSvCDJLUkeaD8Pb22T5PIkW5J8KcmJQ8fZ0No/kGTDUP0VSe5u+1yeJK0+sg9J0mT0npm8H/h0Vf0T4J8C9wEXArdW1Wrg1rYOcDqwur02Ah+GQTAAFwEnAycBFw2Fw4db2937rWv1ufqQJE3A2GGS5FDg14ArAKrqe1X1JLAeuLo1uxp4fVteD1xTA7cBhyU5Cng1cEtV7aiqJ4BbgHVt26FV9fmqKuCaWcca1YckaQJ6zkxeAnwT+B9JvpjkI0kOAZZX1XaA9vOFrf0KYOvQ/ttabU/1bSPq7KEPSdIELOvc90Tg7VV1e5L3s+fLTRlRqzHq85ZkI4PLZCxfvpyZmZl92f2Hlh8MF5ywC2DsY0ybnTt3Lpm57uaclwbnvDh6wmQbsK2qbm/rNzAIk28kOaqqtrdLVY8OtT96aP+VwCOtvnZWfabVV45ozx76+DFVtQnYBLBmzZpau3btqGZ79YFrb+TSuwdv1UNnj3eMaTMzM8O479e0cs5Lg3NeHGNf5qqqrwNbk7yslU4F7gU2A7ufyNoA3NiWNwPntKe6TgGeapeobgZOS3J4u/F+GnBz2/atJKe0p7jOmXWsUX1Ikiag58wE4O3AtUkOBB4E3sIgoK5Pci7wMHBma3sTcAawBXi6taWqdiS5GLijtXtvVe1oy28FrgIOBj7VXgCXzNGHJGkCusKkqv4GWDNi06kj2hZw3hzHuRK4ckT9TuD4EfXHR/UhSZoMPwEvSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknq1h0mSQ5I8sUkn2jrxyS5PckDST6W5MBWP6itb2nbVw0d492tfn+SVw/V17XaliQXDtVH9iFJmoyFODN5B3Df0PofApdV1WrgCeDcVj8XeKKqjgUua+1IchxwFvByYB3woRZQBwAfBE4HjgPe2NruqQ9J0gR0hUmSlcBrgI+09QCvAm5oTa4GXt+W17d12vZTW/v1wHVV9d2q+iqwBTipvbZU1YNV9T3gOmD9XvqQJE3Ass79/xvw+8Dz2voRwJNVtautbwNWtOUVwFaAqtqV5KnWfgVw29Axh/fZOqt+8l76+DFJNgIbAZYvX87MzMy+zxBYfjBccMKgu3GPMW127ty5ZOa6m3NeGpzz4hg7TJK8Fni0qu5KsnZ3eUTT2su2ueqjzpr21P4ni1WbgE0Aa9asqbVr145qtlcfuPZGLr178FY9dPZ4x5g2MzMzjPt+TSvnvDQ458XRc2byq8DrkpwBPAc4lMGZymFJlrUzh5XAI639NuBoYFuSZcDzgR1D9d2G9xlVf2wPfUiSJmDseyZV9e6qWllVqxjcQP9MVZ0NfBZ4Q2u2AbixLW9u67Ttn6mqavWz2tNexwCrgS8AdwCr25NbB7Y+Nrd95upDkjQBi/E5k3cB5yfZwuD+xhWtfgVwRKufD1wIUFX3ANcD9wKfBs6rqh+0s463ATczeFrs+tZ2T31Ikiag9wY8AFU1A8y05QcZPIk1u813gDPn2P99wPtG1G8CbhpRH9mHJGky/AS8JKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSuhkmkqRuhokkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSp29hhkuToJJ9Ncl+Se5K8o9VfkOSWJA+0n4e3epJcnmRLki8lOXHoWBta+weSbBiqvyLJ3W2fy5NkT31Ikiaj58xkF3BBVf0CcApwXpLjgAuBW6tqNXBrWwc4HVjdXhuBD8MgGICLgJOBk4CLhsLhw63t7v3WtfpcfUiSJmDsMKmq7VX11235W8B9wApgPXB1a3Y18Pq2vB64pgZuAw5LchTwauCWqtpRVU8AtwDr2rZDq+rzVVXANbOONaoPSdIELFuIgyRZBfwScDuwvKq2wyBwkrywNVsBbB3abVur7am+bUSdPfQxe1wbGZzZsHz5cmZmZsaa3/KD4YITdgGMfYxps3PnziUz192c89LgnBdHd5gk+RngfwLvrKp/aLc1RjYdUasx6vNWVZuATQBr1qyptWvX7svuP/SBa2/k0rsHb9VDZ493jGkzMzPDuO/XtHLOS4NzXhxdT3MleTaDILm2qv68lb/RLlHRfj7a6tuAo4d2Xwk8spf6yhH1PfUhSZqAnqe5AlwB3FdVfzS0aTOw+4msDcCNQ/Vz2lNdpwBPtUtVNwOnJTm83Xg/Dbi5bftWklNaX+fMOtaoPiRJE9BzmetXgTcBdyf5m1Z7D3AJcH2Sc4GHgTPbtpuAM4AtwNPAWwCqakeSi4E7Wrv3VtWOtvxW4CrgYOBT7cUe+pAkTcDYYVJVf8no+xoAp45oX8B5cxzrSuDKEfU7geNH1B8f1YckaTL8BLwkqZthIknqZphIkroZJpKkboaJJKmbYSJJ6maYSJK6GSaSpG6GiSSpm2EiSepmmEiSui3I/xxrKVl14Sd/uPzQJa+Z4Egk6ZnDMxNJUjfDRJLUzTCRJHUzTCRJ3QwTSVI3w0SS1M0wkSR1M0wkSd380GIHP8AoSQOemUiSuhkmkqRuhokkqZv3TBaI908kPZMM/5101bpDFr0/z0wkSd08M1kEw/8iAM9UJP30M0z2Ay+BSfppZ5jsZ7PPWnYzZCRNs6kNkyTrgPcDBwAfqapLJjykLoaMpGk2lWGS5ADgg8A/B7YBdyTZXFX3TnZkC2+ukOlhQElaaFMZJsBJwJaqehAgyXXAeuCnLkwWw3wC6oITdvHmRQiyZ7J9nbOhrElYjH9gLoRU1aTHsM+SvAFYV1W/3dbfBJxcVW+b1W4jsLGtvgy4f8wujwQeG3PfaeWclwbnvDT0zPnFVfWze2s0rWcmGVH7iVSsqk3Apu7Okjurak3vcaaJc14anPPSsD/mPK0fWtwGHD20vhJ4ZEJjkaQlb1rD5A5gdZJjkhwInAVsnvCYJGnJmsrLXFW1K8nbgJsZPBp8ZVXds4hddl8qm0LOeWlwzkvDos95Km/AS5KeWab1Mpck6RnEMJEkdTNMhiRZl+T+JFuSXDhi+0FJPta2355k1f4f5cKax5zPT3Jvki8luTXJiycxzoW0tzkPtXtDkkoy9Y+RzmfOSf5l+7O+J8mf7e8xLrR5/G7/XJLPJvli+/0+YxLjXChJrkzyaJIvz7E9SS5v78eXkpy4oAOoKl+D+0YHAF8BXgIcCPwtcNysNv8O+OO2fBbwsUmPez/M+TeA57blty6FObd2zwM+B9wGrJn0uPfDn/Nq4IvA4W39hZMe936Y8ybgrW35OOChSY+7c86/BpwIfHmO7WcAn2LwOb1TgNsXsn/PTH7kh1/RUlXfA3Z/Rcuw9cDVbfkG4NQkoz5AOS32Oueq+mxVPd1Wb2PwmZ5pNp8/Z4CLgf8MfGd/Dm6RzGfOvwN8sKqeAKiqR/fzGBfafOZcwKFt+flM+WfVqupzwI49NFkPXFMDtwGHJTlqofo3TH5kBbB1aH1bq41sU1W7gKeAI/bL6BbHfOY87FwG/7KZZnudc5JfAo6uqk/sz4Etovn8Ob8UeGmSv0pyW/tW7mk2nzn/AfBbSbYBNwFv3z9Dm5h9/e99n0zl50wWyXy+omVeX+MyReY9nyS/BawBfn1RR7T49jjnJM8CLgPevL8GtB/M5895GYNLXWsZnH3+nyTHV9WTizy2xTKfOb8RuKqqLk3yK8Cftjn/v8Uf3kQs6t9fnpn8yHy+ouWHbZIsY3BqvKfTyme6eX0tTZLfBP4D8Lqq+u5+Gtti2ducnwccD8wkeYjBteXNU34Tfr6/2zdW1fer6qsMvhR19X4a32KYz5zPBa4HqKrPA89h8IWIP60W9WuoDJMfmc9XtGwGNrTlNwCfqXZna0rtdc7tks+fMAiSab+ODnuZc1U9VVVHVtWqqlrF4D7R66rqzskMd0HM53f7fzF42IIkRzK47PXgfh3lwprPnB8GTgVI8gsMwuSb+3WU+9dm4Jz2VNcpwFNVtX2hDu5lrqbm+IqWJO8F7qyqzcAVDE6FtzA4IzlrciPuN885/xfgZ4CPt2cNHq6q101s0J3mOeefKvOc883AaUnuBX4A/Puqenxyo+4zzzlfAPz3JL/H4HLPm6f5H4dJPsrgMuWR7T7QRcCzAarqjxncFzoD2AI8DbxlQfuf4vdOkvQM4WUuSVI3w0SS1M0wkSR1M0wkSd0ME0lSN8NEktTNMJEkdfv/JeYeKOCFoBQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(genic_scores.shape[0])\n",
    "genic_scores.hist(bins=100)\n",
    "genic_scores.reset_index(inplace=True)\n",
    "genic_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56326\n"
     ]
    }
   ],
   "source": [
    "# select highly conserved motifs\n",
    "genic_scores = genic_scores[genic_scores['scores']>=0.9]\n",
    "print(genic_scores.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>index</th>\n",
       "      <th>scores</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chr10</td>\n",
       "      <td>264199</td>\n",
       "      <td>264204</td>\n",
       "      <td>chr10:264199-264204:+</td>\n",
       "      <td>0.998</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>chr10</td>\n",
       "      <td>282841</td>\n",
       "      <td>282846</td>\n",
       "      <td>chr10:282841-282846:+</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>chr10</td>\n",
       "      <td>294441</td>\n",
       "      <td>294446</td>\n",
       "      <td>chr10:294441-294446:+</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>chr10</td>\n",
       "      <td>299431</td>\n",
       "      <td>299436</td>\n",
       "      <td>chr10:299431-299436:+</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>chr10</td>\n",
       "      <td>300204</td>\n",
       "      <td>300209</td>\n",
       "      <td>chr10:300204-300209:+</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom   start     end                  index  scores strand\n",
       "74   chr10  264199  264204  chr10:264199-264204:+   0.998      +\n",
       "91   chr10  282841  282846  chr10:282841-282846:+   1.000      +\n",
       "99   chr10  294441  294446  chr10:294441-294446:+   1.000      +\n",
       "104  chr10  299431  299436  chr10:299431-299436:+   1.000      +\n",
       "105  chr10  300204  300209  chr10:300204-300209:+   1.000      +"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# format to BED:\n",
    "genic_scores['chrom'] = genic_scores['index'].apply(lambda x: x.split(':')[0])\n",
    "genic_scores['start'] = genic_scores['index'].apply(lambda x: int(x.split(':')[1].split('-')[0]))\n",
    "genic_scores['end'] = genic_scores['index'].apply(lambda x: int(x.split(':')[1].split('-')[1]))\n",
    "genic_scores['strand'] = genic_scores['index'].apply(lambda x: x.split(':')[2])\n",
    "genic_scores = genic_scores[['chrom','start','end','index','scores','strand']]\n",
    "genic_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>index</th>\n",
       "      <th>scores</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1697620</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27859521</td>\n",
       "      <td>27859526</td>\n",
       "      <td>chrY:27859521-27859526:-</td>\n",
       "      <td>0.9138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1697683</th>\n",
       "      <td>chrY</td>\n",
       "      <td>28344541</td>\n",
       "      <td>28344546</td>\n",
       "      <td>chrY:28344541-28344546:-</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1697708</th>\n",
       "      <td>chrY</td>\n",
       "      <td>28438600</td>\n",
       "      <td>28438605</td>\n",
       "      <td>chrY:28438600-28438605:-</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1697764</th>\n",
       "      <td>chrY</td>\n",
       "      <td>28499045</td>\n",
       "      <td>28499050</td>\n",
       "      <td>chrY:28499045-28499050:-</td>\n",
       "      <td>0.9640</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1697810</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358567</td>\n",
       "      <td>59358572</td>\n",
       "      <td>chrY:59358567-59358572:-</td>\n",
       "      <td>0.9702</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end                     index  scores strand\n",
       "1697620  chrY  27859521  27859526  chrY:27859521-27859526:-  0.9138      -\n",
       "1697683  chrY  28344541  28344546  chrY:28344541-28344546:-  1.0000      -\n",
       "1697708  chrY  28438600  28438605  chrY:28438600-28438605:-  0.9336      -\n",
       "1697764  chrY  28499045  28499050  chrY:28499045-28499050:-  0.9640      -\n",
       "1697810  chrY  59358567  59358572  chrY:59358567-59358572:-  0.9702      -"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genic_scores.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "genic_scores.to_csv(os.path.join(output_dir, 'GCATG_RBFOX_MOTIF_POSITIONS_ALL_GENIC.phastCons46way.conserved.bed'), sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3-essential",
   "language": "python",
   "name": "python3-essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
