{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# In this analysis, we are taking a set of peaks and finding those peaks that contain a motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import pybedtools\n",
    "import pysam\n",
    "import re\n",
    "import urllib\n",
    "import sqlite3\n",
    "import random\n",
    "from random import sample \n",
    "from collections import Counter\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "from xml.dom import minidom\n",
    "from Bio import SeqIO\n",
    "from matplotlib_venn import venn2\n",
    "pd.set_option('display.max_colwidth', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/01_SAILOR_bulk_rnaseq/inputs/'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/01_SAILOR_bulk_rnaseq/outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "slop = 0  # leave at 0 since we're starting with the full 3'utr region.\n",
    "motif = 'TTTTC'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define our peak sites\n",
    "- These may come from multiple sources. I'm trying to classify them according to how they were annotated (if they were annotated). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_genes = '/home/rmarina/projects/apobec/ref_files/histone_genes.bed'\n",
    "regions_of_interest = '/projects/ps-yeolab3/bay001/annotations/hg19/gencode_v19/hg19_v19_cds_and_three_prime_utrs.bed'\n",
    "\n",
    "assert os.path.exists(hist_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in hist genes and intersect geneids with our regions of interest to get the CDS and 3'UTRs for these genes only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of histone genes: (should be 122 but in case Ryan updates this file) 122\n",
      "Number of histone genes that are missing CDS or 3UTR: 33\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000203818.6</td>\n",
       "      <td>HIST2H3PS2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>149400134.0</td>\n",
       "      <td>149400542.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000203818.6</td>\n",
       "      <td>HIST2H3PS2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>149400130.0</td>\n",
       "      <td>149400134.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>HIST2H2BF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>149754791.0</td>\n",
       "      <td>149754816.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>HIST2H2BF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>149765522.0</td>\n",
       "      <td>149765643.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>HIST2H2BF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>149783500.0</td>\n",
       "      <td>149783878.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneid    genename chrom        start          end  score strand\n",
       "3  ENSG00000203818.6  HIST2H3PS2  chr1  149400134.0  149400542.0    0.0      -\n",
       "4  ENSG00000203818.6  HIST2H3PS2  chr1  149400130.0  149400134.0    0.0      -\n",
       "5  ENSG00000203814.5   HIST2H2BF  chr1  149754791.0  149754816.0    0.0      -\n",
       "6  ENSG00000203814.5   HIST2H2BF  chr1  149765522.0  149765643.0    0.0      -\n",
       "7  ENSG00000203814.5   HIST2H2BF  chr1  149783500.0  149783878.0    0.0      -"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes = pd.read_csv(hist_genes, sep='\\t', names=['chrom','start','end','geneid','genename','strand'])[['geneid','genename']]\n",
    "print(\"Number of histone genes: (should be 122 but in case Ryan updates this file) {}\".format(all_genes.shape[0]))\n",
    "all_regions = pd.read_csv(regions_of_interest, sep='\\t', names=['chrom','start','end','geneid','score','strand'])\n",
    "\n",
    "regions = pd.merge(all_genes, all_regions, how='left', left_on=['geneid'], right_on=['geneid'])\n",
    "print(\"Number of histone genes that are missing CDS or 3UTR: {}\".format(regions.shape[0] - regions.dropna().shape[0]))\n",
    "regions.dropna(inplace=True)\n",
    "regions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Re-format these 243 regions (exons) into a BED-like format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "243\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>geneid</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149400134</td>\n",
       "      <td>149400542</td>\n",
       "      <td>ENSG00000203818.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149400130</td>\n",
       "      <td>149400134</td>\n",
       "      <td>ENSG00000203818.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149754791</td>\n",
       "      <td>149754816</td>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149765522</td>\n",
       "      <td>149765643</td>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149783500</td>\n",
       "      <td>149783878</td>\n",
       "      <td>ENSG00000203814.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start        end             geneid  score strand\n",
       "3  chr1  149400134  149400542  ENSG00000203818.6    0.0      -\n",
       "4  chr1  149400130  149400134  ENSG00000203818.6    0.0      -\n",
       "5  chr1  149754791  149754816  ENSG00000203814.5    0.0      -\n",
       "6  chr1  149765522  149765643  ENSG00000203814.5    0.0      -\n",
       "7  chr1  149783500  149783878  ENSG00000203814.5    0.0      -"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions['start'] = regions['start'].astype(int)\n",
    "regions['end'] = regions['end'].astype(int)\n",
    "regions = regions[['chrom','start','end','geneid','score','strand']]\n",
    "print(regions.shape[0])\n",
    "regions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For this dataset, we already know that these cover 3'UTRs and CDS regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get the fasta sequences and check for existence of motif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert our filtered peaks to BedTool objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "size prior to merge: 243\n",
      "size after merge: 142\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bay001/anaconda2/envs/python3-essential/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n",
      "/home/bay001/anaconda2/envs/python3-essential/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "genome_fasta = '/projects/ps-yeolab3/bay001/annotations/hg19/hg19.fa'\n",
    "chrom_sizes = '/projects/ps-yeolab3/bay001/annotations/hg19/hg19.chrom.sizes'\n",
    "\n",
    "def get_peak_interval_from_annotated_dataframe(df, slop, chrom_sizes=None):\n",
    "    \"\"\" \n",
    "    Returns a bedtool from a peaks dataframe. \n",
    "    Peaks dataframe is generated from above (regions).\n",
    "    Merges the bedtool so we reduce chance of overlapping. \n",
    "    Merging also \"connects\" CDS and 3'UTR regions\n",
    "    \"\"\"\n",
    "    \n",
    "    bedtool = pybedtools.BedTool.from_dataframe(df).sort()\n",
    "    bedtool = bedtool.slop(b=slop, g=chrom_sizes)\n",
    "    print(\"size prior to merge: {}\".format(bedtool.to_dataframe().shape[0]))\n",
    "    bedtool = bedtool.merge(c=(4,5,6), s=True, o=('distinct','collapse','distinct'))\n",
    "    print(\"size after merge: {}\".format(bedtool.to_dataframe().shape[0]))\n",
    "    return bedtool\n",
    "\n",
    "# utr3_bedtool is a bedtool-ified version of the annotated dataframe for JUST 3'UTR peaks.\n",
    "peaks_bedtool = get_peak_interval_from_annotated_dataframe(regions, slop=slop, chrom_sizes=chrom_sizes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Need some kind of set of genes that are considered \"bound by SLBP.\" \n",
    "- So we're going to use the original gene list from Ryan and return the genes that have a 3'UTR or CDS (so not Pseudogenes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000178458.5',\n",
       " 'ENSG00000178762.3',\n",
       " 'ENSG00000181201.2',\n",
       " 'ENSG00000203819.6',\n",
       " 'ENSG00000206417.4',\n",
       " 'ENSG00000213244.3',\n",
       " 'ENSG00000213440.2',\n",
       " 'ENSG00000216331.1',\n",
       " 'ENSG00000216436.2',\n",
       " 'ENSG00000217646.1',\n",
       " 'ENSG00000217862.2',\n",
       " 'ENSG00000218281.1',\n",
       " 'ENSG00000218502.3',\n",
       " 'ENSG00000218690.1',\n",
       " 'ENSG00000220323.3',\n",
       " 'ENSG00000220875.1',\n",
       " 'ENSG00000223345.3',\n",
       " 'ENSG00000223383.1',\n",
       " 'ENSG00000224447.1',\n",
       " 'ENSG00000226881.4',\n",
       " 'ENSG00000226908.1',\n",
       " 'ENSG00000230618.3',\n",
       " 'ENSG00000233320.2',\n",
       " 'ENSG00000233733.1',\n",
       " 'ENSG00000234289.4',\n",
       " 'ENSG00000234612.1',\n",
       " 'ENSG00000235655.2',\n",
       " 'ENSG00000236534.1',\n",
       " 'ENSG00000240929.2',\n",
       " 'ENSG00000242387.1',\n",
       " 'ENSG00000255329.1',\n",
       " 'ENSG00000259389.2',\n",
       " 'ENSG00000265541.1'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQ0AAADtCAYAAAClISA+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3Xl8W+Wd7/HPT9Zmx3acxNnJHsKSAgHCVtYSKEsLzJ3bQjfovgBDOy1T2umUqmpLp7R3SpfbhYHetnM7bWnpQMtSoExJCGENe4AkkD04ibPZsWNJ1vKbP85x4rheJFs6R7J+79dLr9iWdJ7HjvTVc57tiKpijDH5CvhdAWNMZbHQMMYUxELDGFMQCw1jTEEsNIwxBbHQMMYUxELDGFMQCw1jTEEsNIwxBbHQMMYUxELDGFMQCw1jTEEsNIwxBbHQMMYUxELDGFMQCw1jTEEsNIwxBbHQMMYUxELDGFMQCw1jTEEsNIwxBbHQMMYUJOh3BUzlk7iEgQgQ7fVvFOf1pe4N998skOl1SwCdGtOEx9U2wyR23RMzFIlLAGgCxru3cUAjB0OiGC3WLNAJdLj/dvb6fo/GNFWEMkwRWGiYQ0hcIsBUDgbEeJyA8PtUtg1odW87cILEXrw+sNCocm4rYhJwmHubCIivlcpPGtiJEyLbgRaNacbfKlUHC40qJHFp5GBITAPC/taoKDLAZmADsFljmva5PqOWhUaVkLjUA4e7tyafq1NqWeBNYD2wyfpDistCYxSTuASB+ThBMdXn6vglB2wDXgfWaUyzPten4llojEISlwnAUThhEfK5OuUkCawBXtWYdvhdmUploTGKSFxmAcfjdGyagSlO/8dLGtNtflem0lhojAISl9nACUCzvzWpSK3AS8AGG8LNj4VGBZO4zMEJiwl+12UUaAOe1phu9Lsi5c5CowK5YXEizsQrU1zbgCc1pjv9rki5stCoIBKXycAZWMvCC2/gtDw6/a5IubHQqADugrBTcEZEjHeywMvACxrTbr8rUy4sNMqcxGU+cBpQW4rjT0yTnpskMztFbkIGacxCfQ6pzyJjckhtjkBtjmBYkRygoArkBO0WcmnnX20Pkt0RQreHYFsIWsIEtoUJ7gmOipXUSZxTlrV+V6QcWGiUKXeq9xk4U71HLJojt7CL7vlJsnNSyIxuApPThCJKTTGOP5AM5NqCZN6IkllVCy/VEdwYIaRSEetb+toIPKoxTfpdET9ZaJQZdwHZcTjzLYb9KV2j6NEJUid3kl20n5oZ3URqymQhWkrIborQvTpK7oUxBF6qI5IK+L6KNl8JnODY5HdF/GKhUUbc9SHnMczJWZO7SZ/RQfqk/cj8JOFStyKKJQ25VXUkH2mExxuIVkiArAEer8aFcRYaZULiMgN4G87GNnlrzJBdso/Uee0EZnYX9txylIbc6lpSjzSiKxqIdNWUdfB1AEurbVaphYbPJC6CM+fiePI8fRBFF+8nedleeEsX0XI57Si2DOReriN553hqXhpDxO/6DECBF4FnqmVGqYWGjyQuUWAJMD2fx0dz5N69m8QF7YTHZqtrIdqOIN33jCNzfxPRdHmevmwE/loNGwFZaPjEnah1HjBmqMfWZclesZvkxW1EoxXST1EqXUL2gSaSfxhPdF+w7P4Wu4EHNKb7/a5IKVlo+EDichRwOkPsu1mfJfu+XaTe3k6kUjo1vZKG3F+aSPyymWiZ9Xt0AQ+O5mnoFhoek7gsxllkNqDGDNkP7CK1ZB/RsJZlU7xsdAnZ3zaT+tM4arPlM/cjg9NBut7vipSChYZH3A7PMxhiKvg795K4cifhOmtZFGRXkO5/n0T2iYbSzJwdppUa0+f8rkSxWWh4wJ2wtQSYM9Bj5iTp/tw2crNHwbCpn9ZFSP5wCrIuWjajLauB5aNpZMVCo8QkLjXA+cDM/u6P5Mh9rJXE+e3UjdahU6/lQO9vouv2SdSVySnLGpxZpKPizWahUULuxr4X4lwm4G+c1kHi2h0Eq2341CstIVLfnI5sipTFJRrWAstGQ3BYaJSIGxgXA1P63lej6HXbSSzZR533NasuGcj9dgKJO5qHHtr2wFqN6VK/KzFSFhol4HZ6ng/M7nvfxDTp+FZyM7rL5py7KqyLkLxpOjU7Q7636l7RmK7wuQ4jYsN5pXE6/QTGaR0kfryBgAWG9+aliP5kA4HTOvD76vQLJS4n+1yHEbGWRpFJXBYBh7woahT95A4SF7Xb6YjfcqC/aibx+wm+/188pTF90ec6DIuFRhG5u2yd2/tn9VmyN20hPTdlQ6nlZFkDXbdM9X1C2IOVuC+HhUaRSFym4XR8Hjjlm5gm/e3NaHOmLHrvTR9royRuPIywj9PQu4G7NaZtPpU/LNanUQQSl/HA2+n195yTpPsHGxELjPK1IEntDzeSndyNXxvphIG3uxtHVwwLjRFy/8MvhIPhcFQXqe9spqY+Nyo21R3VJmUIf38TMieJX7uNN+FsvlQxLDRG7iygvueb4/eT/MbW0m/Ya4pnTI7gtzZT42NwzHIXMlYEC40RkLgcCczt+f6kThI3biVsK1MrT51S43NwnOBek7fs2Yt7mCQuTcBbe74/IkHqiy1EQvY3rVh1Ss2/biEwI+VbcLzNfV2VNXuBD4O7avVc3EsMTO0mHd9KjbUwKt+YHMGbNxOYmPalczQEnO3OKC5b9iIfnlOAZnA2zLl5s/Ni87lOpkgacgT/bRM0ZfBjv8/JwEIfys2bhUaB3EsNHAPOsvZvbSEzzlapjjrjsoTiW8nUKH5MZDrZvcJeWbLQKIDEJQKcA87U8K9tIWXrSEavuSmi1233Za1KEGdUrixZaBRmMe6FmK/bTuLoZFltLWdKYMk+6i7aS5cPRU9zN6AuOxYaeXJnfR4NcHqH7YVRTT7RSvSIBCkfij5F4lIO+4AcwkIjf28FpDlN5h+32dTwahKEQGwrNT50jIaBMz0uc0gWGnmQuMwFpomisa1kqv2CRdWoIUcwvpWMeN8xOlPiMsvjMgdloTEEd9u+UwE+upOE7RZeveamiF6+25f+jZPKae6GhcbQFgH1x+8necle6/isdlfspnaq96tixwPzPC5zQBYag5C4NADHRXPkPt9CMGCXGKh6IQh8oYWsD0Uvdmci+64sKlHGFgM1H28l0WAzPo1rXoroZXs8P01pBI70uMx+WWgMwJ2RN39Giu4ltren6ePKXUR8WJ9ygtvH5isLjYEtAuSz28jZlc9MXxGl5oYWz4dg6yiDdSkWGv1wJ9QsOKedxOG2IbAZwJFJan24JMIiv7cHtNDo3zGRHHxsp/VjmMF9ZKfn76EIPvdtWGj0IXEJAUdetZOEXWPVDGVKmsi57Z63Nhb6OW/DQuNvHdGUIXBRm83JMPm5aic1Hi+hbwBmeljeISw0enHT+5j37yJl2/aZfE3IEr64zfPWxls8Lu8Ae2McakZjhrpz91krwxTmit2EQjlyHhY5XeIy1sPyDrDQONT8K3aTtL0+TaHGZgn93V6SHhd7hMflARYaB0hcgpEcM89rtyFWMzyX7iXk8SrYBX5MLbfQOGjWxW1k6mzZuxmmpiyht3Z62tqoA2Z4WB5godHbvMv22LwMMzJ/v8fzIg/3ukALDZwNg0/Yz6QJWduRy4zMgiS107292NJhXp+iWGg45ry9ze8qmNHi0jZPF7KFgakelmehARDJMW/xfrsUgSmOs/YR8bhD1NPtAKs+NCQudad1Ms6u8m6KpT5H0OMOUU9nh1Z9aADTzmu3pe+muM7Z52lLo9HLC0dXfWiMzzBpYZedmpjiOqaL8Gg9Ran60Liwjeag/R1MkY3JETwy4ekoioWGFyQutcd24cv8fTP6ndXh6c5ek91rDZdcVYdGKMeU+Umbm2FKY/F+T/djEWCiFwVVdWi8tZNJNmpiSmVKmrDHmw9baJTawi4m+V0HM7qd0eFpaDR7UUjVhobEJTIrhWfDVKY6Hb/f0+IsNEopnGPivJT1Z5jSmp3ytF+jwYvO0KoNjUv30mj9GabUxmUJ1WU9vYxjyVsbVRsap3TS6HcdTHVYkPR06NVCo1TqsxYaxhtHJqylMSpE1SZ1GW8sSHq6tmlCqQuoytCQuASiOer9roepDrNTnu4IN6bUBVRlaCzaT319zq6eZrwxIUPQw8sbhEp9ZfmqDI2L2+zUxHgnADI+42m/Rl0pD16VoTExbZ2gxlvNFhqVrTlT+vM+Y3qbkPF0bw0LjWIbk7ULIhlvWWhUOMF26jLeak5baFS0ALbmxHhrQsbTuRolDY2qu6KYiFxYG+TfRQgdM4nHvnwWD/hdp3J182MseXEHZwI6vpY3v7mEXzy0jnl3r+ZdOSU4sY5N3z6f/4gEPb1aekUal09odBHkx3yeHEGUGqbzLB/gHm7lKtqZBQh17OBKfsFYUoMcqbZoFe+HqHrZavKXiNQAa//lTL53zCQyn7iXL129mNvPmMk2v+tWbl7bSVNsKTfcdgmxsVHS197HJ45sZtWjm7n0n07ju6ccRutXHuHS5jp2f/oUVvhd33K3OUzy2jlD9KXlgA4ijCVFihp+yOc5mzuYzzbGuZdE+DnvJkoH7x30w+5Njel9xav9oart9ORk4I1Tp7N7TJjsUc08s3Qjx/ldqXKlENiXIpTKEEjnCEeDdAeEzCmH0Qpw4lRefbmVE/yuZyUQ8mhpBOBACyJNDT2rsHsCIwdkCCFD9o+U9H1dbacn04EtotSowPha9m5oY67flSpHR02k7aRpPHTdn/lWTYD0tAZe/egJrPzLev73Q+uY9fZ5bFqxhRO7uhnnd10rQU2+DfoMwnf5MkkmMpOlnMQGAG7lg+zkGMbQwnu5c4ijlLT/pNpaGtLnX8DTXu2KsXUfdat3seiWC/jSr/4XN6SzhG9dySkfPI7bfv0yl191F/8cCZIUsf6MfNRonm/kIMoNfJ1r+QJ7mc0qpgHwSX7JF/k8DWznrywe4ijW0iiircCMnm/2JBg3NoJd+rkfD7zBUY0Rds0ZRyfAMZN5ft1e5l19Ek+9YwHfAbhjFUfv6WKyvzWtDFroZ/8EEkxkLa+wkLfQAjiBcjTPsJILgMeLXsk8VVtL4xng8GdaaN7fTc1ruzjp7Nm86HelytH0Bva07mfu3gThnMLaXRw5uZ5t6/bSANDZTfCh9Vzwtjks87uulSCXT4t2O/Xsdkc+9hOilaNoZgdr3V3Gc8AajqOB7UMXVzpV1dJQ1YyI/MN3nuRnQM3Ciaw400ZO+nXR4Wx48k2eveZ+/kUg11zHlqsXs/ym5Vy2sY1jVZHjp7LsXUezxu+6VoJsPi2NVsbyIB9GCaAIh7GSc3iZ7/F5skRRhEa28h7+c4gjlTQ0qmrItce975PLwXYiN97ZGCZ53VBDrsVjQ64l4OU1No2ho8bTDveSrqitytDIWWgYj+0JehoaXaU8eLWGxmBTcI0put0WGpVNLTSMx3aFPF2wZqFRbO1Bd1quMR7ZFbTQqGi7g86EJWO8sivo6XvNQqPYtoZp97sOprrsDnl6CVALjWL7r/Hsy2uGnjFFkIHc3hpPQyNRyoNXZWhsjtDVGbBhV+ONbWEyKp71aaQ0pjZPo9g0ppoKsM/vepjqsCHi6QWgO0pdQFWGBkDSQsN4ZE3U05GTXaUuoGpDY0/QOkONN9bUetqfYaFRKksbS9+MMyYLuj7i6WpyC41SeXgs7R0BT881TRXaGaI7HfDsfZYDdpe6kKoNDWDn61GbTm5Ka0PE02u4tpV65ASqODQ0ptnNkdKnsqluT9V7WlzJT02gikMD4Lkxzlb8xpRCFvSJek8vAWqhUWrPj6HV+jVMqWyM0N3l7UxQCw0P7Fhda/0apjSeqvf0AykD7PSioKoODY1p+rkx7PW7HmZ0eqzB0wuNb/WiExSqPDQAHhrLrqR42sNtqsDuGrq3RAh5WORmrwqq+tDoDrDjmXo7RTHF9dwY0h4XaaHhoZYHx9oyeVNcfxnr6SzQnRrTku6h0VvVh4bGNP3iGLa313j+yWBGqdYg3a/VeTrU6lkrAyw0eqxb0WD7a5jieGis5x9Am7wszELDseWBsXb1czNyGcj9ucmzK6kBdGlMPZmf0cNCA2dK+YYoLRvDtku5GZlnx5DaF/R0QtdGD8sCLDR6e+N3E6xD1IzM7yd4/p5a63F5Fhq9tDzWgO6usb4NMzxbwqTW1HraAbpXY+r5+ikLDZfGVFVYf9d4W4tihuc3EzzvF1vtcXmAhUZfr987jlpbxGYK9WaI1PJGaj0sMge87mF5B1ho9KIx3ZkVdt413k5RTGFun+R5f9g6jakvHfcWGn/rpbvHEe201obJ07oIyZX1ng6zAqzyuLwDLDT+1oZ0gK5fNVtrw+TntkmeF9mqMfVkGXx/LDT60JgqsOq+cdRtCdtCNjO4V2pJvlJXPa0MsNAYyGtA6pYpYNd8NQPJgd46yfP3UDuw3uMyD2Gh0Q+NaRp4+fVaIisaSnsxXVO5HmkksSHq6UY7ACs1pr4uebDQGNgrQPdPJxGxTXpMX3trSP9ksuenJbs0pus8LvNvWGgMQGOaAl7ZF6Tm183Wt2EOdctUsinvLoLUY6XH5fXLQmNwLwBdd4+j1jpFTY8V9XQ9P8bzVsZ2jamn+2YMxEJjEG7fxtMqSHw6gZSdplS9jgCZH0zxdH1Jj6d9KLNfFhpD0JiuBVp3hAn9aLK1NqrdD6eQ9vhaJgBbNKbbPS5zQBYa+VkB8MhY6lbU49lejKa8LG+g64kGT9eXgDPk/4zHZQ7KQiMP7uy7NQC3TCXaGrTZotVmc5jkv031PDAAXvF6Z66hWGjk72mgOxUg8I3pzrZuflfIeKMzQObGGQSzgnhc9D7KqC+jh4VGnjSmCeA5gA1Rwv9vkm0NWA2yoF+bTnZP0NNLEvR4VGNadgsnLTQKswrYDXDPOOrua7L+jdHuZ5NIeHw5gh6vaUxbfCh3SBYaBXCn7/43zsV2+elk6p60jtFRa1kDXfeMo86HojuBp3woNy8WGgXSmLYBj/d8/61p1K6O2vqU0WZ1lMQt/nR8AizXmJZtZ7uFxjBoTFcDGwCygtw4g8ibIZvDMVqsj5D88gwiPnR8AqzVmG7xody8WWgM36M4zUiSAQJfnEnNXru0Y8VrCZH65xmEfFhXAs5oyRM+lFsQC41hche0/RV3v422IMEvzAS7JmzlagmRun4WQR9mfILTT/aQ+7oqaxYaI+BO7X2+5/ttYUKfnQV27ZTK0xMYnf4EBsBSjeken8ouiIXGyD0LbOv5ZmeI0D/OJrDDZo1WjC1hUp/1NzBe0Jj6uhtXISw0RsjdU/QhoK3nZ21Bgp+eTc36iE0AK3fP19H1mVmEfDolAdhMma0tGYqo2haYxSBxaQAug4Pj+qEcudibpI7r8m3ozgziT03sv20yY3ysQjtwVzkPr/bHQqOIJC7NwCVA6MDPFL1mB10Xtvv64jS9ZCD308kkH2zyZeJWjzROYLQN+cgyY6FRZBKXGcAF9Dn1O3Mfic9sJxxR35rBBugSst84jPTL3l92oLcc8GC5z8cYiIVGCUhcjgTO6vvzyd2k41vJTU/7spah6rWESH31MALbwgdbgj5Q4L8rqeOzLwuNEpG4LAZO6PvzUI7cp7eTPKfD16ZxVcmB3t9E1+2TqPNplmdvS93d4CqWhUYJSVxOBxb2d9+Sdrqu2UE0rDaCVUp7a0jfPI2sD1dB689jGtNX/a7ESFlolJjE5VTg2P7um5gmff02MgsTNrpSCssb6Pr+FKI+TQnva1QEBlhoeGKgU5UeZ+wj+alWasZmfT3XHjU6nR3D0z7s5zmQ5RrT1/yuRLFYaHhE4nIccMpA90dy5D7WSuL8dupq8P28uyKlIffnJhL/fyK1yfJoXShOYKz2uyLFZKHhIYnLApxRlQFf0HOSdH9uG7nZ3WVxDl4xnqyn69ZJhHeFfNmWrz/dwMMa061+V6TYLDQ8JnGZCZwHg7+4T+8gceVOAjY8O7h1EZI/moy8XltWf6d9wAOVOHErHxYaPpC4TALOh6FniZ7SQeKDuwjM6C6rN4XvtodI/WIiuRXl02/RYzvOEvdRu+7IQsMnEpcosASYns/jT+wk+aGdUO2nLWujJH49AXm2viz/DmtxdhAf1Ze3sNDwkcRFgBMZZGSlr+P2k/z7PeSO66K2WjpM05B7soHk78cT3BAl7Hd9BvC0xvQFvyvhBQuNMuCuV3kb5P/p2Zghe3EbqfPbCU7KlO0baURaQqQeHkv2viYiPi5dH0oSWKYx3eR3RbxioVEmJC71OB2kkwp97rwkqYvbyJzWQaQhVzajB8OyJUxqeQOZvzYS3uHvGpF8bMI5Hamq3egtNMqIxCUAnMYAU8/zcXiC1KmdZI7fT82cFOFgmW+0lAPdHCb1aCO5RxrLash0MGngidE2/yJfFhplSOIyDTgDaBrJcSI5csd2kTq5k9yiLoJTymD4NiVkN0XofrWW3Et1BF6pJVzGpx792Q48ojHt8LsifrHQKFNuq+N4YBEU500VzZGbkyQ9N0V2bgqdmSIwNU2wFNPXs6AdNWR2B8ltD5F9pQ5eqiO4OUxI/V9pOhxZnG35Xna3eKxaFhplTuIyFjgTmFaqMuqyZGelyIzPkmvMoGOz0Oje6nNIQxaiOSQH5MSZG50WNBWARABNCewOottDSEuYQEuYml1BghUaDv3ZBqyolN3CS81Co0K4U9BPpYARFjNibcBT1TQykg8LjQoicYkAJwFHUuYdnBUuCawEVo/2iVrDYaFRgdzh2UVYeBRbBliFcx2Sitoh3EsWGhXMwqNocsAbwEqNaafflSl3FhqjgMRlDAfDo5KGL/2WBlbjjIhYWOTJQmMUccPjWGAB+D8no4x1Aq8Cr9ppSOEsNEYhiUsNMBc4Cpjic3XKyVacsNhU7XMtRsJCY5STuDThtDzmA/U+V8cPe4ANwBsa03a/KzMaWGhUEXd6+nxgFpTd5jXFtBMnKDZYUBSfhUaVkrhMAA5zb1Oo7A5UBXZwMCisU7OE8lpRKCLvAT4LvAXYj/Of80vgJ1qBqSMi7wPeqarvE5Ew8CXg/ThTtXcCfwW+pqobRWQpzkzMNM6L83Xg98Atqppyj9cEfBe4GGcLv23Az1T15hLV/xGc/4sIzv/FV1T1j+59XwXmq+oH+jxHgcOBP+K0NMBpbaRx5icIU/kPmsnxMlcToBtQouxiMXdzLi+X4ncZpjTQ6t52AK2jeXu9cjNkaIjI9cANwLXAgzg9z4uAfwJ+BqRKWcESuRi43/36TpxP2/cBz+O86T+AsxXfz9zH/IOq3i4iY3BmZH4POF9EznND8xb3eUcB7Th9CG8ZaSVFJKiqmX7u+gzwqqpmROQU4GERWaCq24Y6pqoeWHbvBuKvVPX2Xj/7EMIivsLldDOe3/BBlnMtx3I9zfRXl1JTnL/pDvfWCuy1jkz/DHp6IiJjgRbgKlX9wyCPiwA3AZfjfPrdBXxWVRMicg7wK5w31hdwVgt+SVV/nsdzm4Ff4CwTzwGvAGerHjq1V0TiwHhVvU5EQjhrBn6kqjeISC2wF5iqqntFJIDTEliIE373AAtU+7+C9wBvrJk44/uXq+q9IrIK+LKq3j3gH/PQY14K/CvO/qAvAFerOhfTEZGNwE9wWj5HAGMGCI6eY50MPAqcpapPD9XSUNU3hvjdPgR8TFXPcL+vw2ldnsxXeQ0Y794acdbBRNxb1L0VOsksh/NB1Al09Pm3E+i0qdzlZaiWxmk4L4g/DvG4m3GG+BbhNB1/DXwF+Gf3/inAWJw3yfnAnSJyt6ruHeK51+MMk010j3MqzidPX8uA77tfn4Sz58HZvX6HNW5ZACcD61V1l4icBzw9UGAMRFU3i8hKnNWn9wJPAjeJyDjgMVV9faDnisgC4DfA3wFLcU777hGRo1UPzBl4L/AOYNdAgSEi9+Ls9BXBaQGuLOR3yIeI1AAfxvl/2eT2FXQCmwd8TlxCHAyRnteX9rplcU6HMkDG5klUIFUd8IbTTN/e52eP43ySJ3Au/CM4n0Tzej3mNGCD+/U57mODve5vxQmAoZ77NZzAmj9EPWtxFhlNAL6I00exFWeIMQ78oNdjvw7c6H59G/DbIY69FOeTt+/Pfwvc1qv8LwHP4rzB3gAuGuB4NwK/6/V9AHgTOMf9fiPwkcHq1Ou5IeAinJZZz8++itN66PtY7ft37O93Az6E84Zuc3+XBE6Lasj62K06bkM1JXcDzSJyoEWiqm9V1Sb3vgBOK6AOeFZE2kSkDXiAg60DgN166Cdml/uGHuq533HfgA+JyHoR+WJ/lVTVBM4n7dk4QbYMJ9xOd3+2rNfDe/dn7AamDvE3GMh0nDkAqGpCVb+pqifiBNfvgN+LyPh+njcNZ2/JnrrngC0ceimDvFo+qppW1T8DF7inPOC84Q/ZVMc9ZQMnBPLxpPt/PA74E06Lyhhg6PPPJ3A6Oi8b5DG7cD6NFqpqk3sbq6r5TCQa9Lmq2qGq16vqXOAS4HMismSAYy0DzsXZ7eoZ9/sLcE5HHgUQkSk4IfGc+5yHgZNF5LA86nqAiMzAufTA8r73qeo+4Js4HaNz+nl6CwdHLxARAWbgtDYOHKaQ+uCcBsxzv94MzO5z/xyc04I3KYCqdgLXAFeKyPEF1smMUoOGhqq24TTvfywi7xKRehEJiMgi3KuDuZ+UtwG3iMgkABGZLiIXDFX4UM8VkXeKyHz3jbUP54WfHeBwy4CrcEYVunGb3jinOjvdx1wMPKCq6pb/MPAX4C4ROVFEgiLSICKfEpGP9C1AROpE5GycU6ancVssInKjiJwkImERieKMbrQBa/qp5++Ad4jIErcFcD1OMD8+1N/LLetIEblIRGpFJCQiH+Bg6wqcltoRInKle/94nBC7UwfpUB2Iqu4GbsfpZzJm6J5uVf028DmcYdeecfFbcUZCel7oX8A5jXhSRPbhfIIfkWcdBnvu4e73nTitnh+r6tIBjvM4Tt/Co+73r+L0czza6zG9T016vMv92R04Q3urgMVuuT3+r4h04Pzu3wP+AFyoB0dxFPg5TsupBaez9x3uJ/UhVHUNTl/NtwkMAAACnUlEQVTRD93HXwJcopp3h6Dg9Fu04swp+Qxwhao+5x6/1f09P+k+ZpX7e12d5/H78z3gYhE5dgTHMKNE1cwIdftltuN0utrUYmOGqZo2bhmPM2pigWHMCFRNS8MYUxzV1NIwZkgi0ikic4fxvNkioj3TE0TkzyLyweLX0H8WGsYzIvIeEXlKRPaLSKv79TXu6FhZUNV6VV1fhONcpKq/HOpxbtDMH2l5XrLQMJ5wFz5+H2fC3hRgMvApnAl4o/Kq96OVhYYpOXfh49eAa1T1TnfSnqrq86r6fj24xUBERP6PiGwWkR0i8lN3wSEico6IbBWR691WyjYR+XCvMgZ7brOI3OvOOt4jIsvdhYv91fXAJ7+I/EJEfiQi94lIh9symtff8/o5zlIR+Zj79XwRWSYi7SKyS0TucH/eMx3gRfe06Iph/YE9ZqFhvFDIwscFOIsX5+NMre89qaz3wsePAj9yFwkO9dzeCx8n46wTyncE4L04ExzH4cwnuinP5/X2deAh9xiH4czRQVXPcu8/zj0tumMYx/achYbxQjN9VuyKyOPuJ39CRM5y+zU+jrP4bo+qduDMZH1Pr+OkcTZHSqvq/TiT/o7I47lpnOUDs9znLtf8hw3/S1Wfduv+nzihVKg0ztKBaaqaVNXHhnGMsmGhYbxQEQsfB7C9n/IKdQPOTN6nReSV/pYoVBILDeOFSlr4WHSqul1VP66q03Cm9/+40kZMerPQMCVXYQsfi05E3i0HV1Lv5eBmROCsZyp4XoifLDSMJypo4WMpnAQ8JSKdOPuTfEZVN7j3fRX4pXtadbmHdRo2m0ZujCmItTSMMQWx0DDGFMRCwxhTEAsNY0xBLDSMMQWx0DDGFMRCwxhTEAsNY0xBLDSMMQX5H+1ntKUnUwOkAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x = peaks_bedtool.to_dataframe()\n",
    "venn2([set(x['name']), set(all_genes['geneid'])], ['Genes w/CDS or 3UTR', 'Genes in list'])\n",
    "set(all_genes['geneid']).difference(set(x['name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(output_dir, 'histone_genes_with_cds_or_3utr_annotations.txt'), 'w') as o:\n",
    "    for i in set(x['name']):\n",
    "        o.write(\"{}\\n\".format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def record_id_to_coords(id_string):\n",
    "    \"\"\" \n",
    "    Turns a seqrecord fasta id into coordinates.\n",
    "    seqrecord fasta ids (stranded) look like this:\n",
    "    \n",
    "    chr11:1774028-1774186(-)\n",
    "    \n",
    "    \"\"\"\n",
    "    chrom, pos = id_string.split(':')\n",
    "    coord, strand = pos.split('(')\n",
    "    strand = strand.replace(')','')\n",
    "    start, end = coord.split('-')\n",
    "    try:\n",
    "        assert (strand == '+') or (strand == '-')\n",
    "    except AssertionError:\n",
    "        print(\"Strand not correct: [{}]\".format(strand))\n",
    "        return 1\n",
    "    return chrom, int(start), int(end), strand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the sequence from the 3'UTR peaks and save them. This will report strand-specific, so negative stranded coordinates will return the reverse complement. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_peaks_as_fasta(bedtool, genome_fa, output_fasta):\n",
    "    \"\"\"\n",
    "    Writes peaks as strand-specific FASTA sequences. \n",
    "    s=True means we will be returning the reverse complement \n",
    "    sequence for negative strand peaks.\n",
    "    \"\"\"\n",
    "    seq_bedtool = bedtool.sequence(fi=genome_fa, s=True)\n",
    "    with open(output_fasta, 'w') as f:\n",
    "        f.write(open(seq_bedtool.seqfn).read())\n",
    "\n",
    "output_fasta = os.path.join(output_dir, os.path.splitext(os.path.basename(hist_genes))[0] + \".fasta\")\n",
    "\n",
    "write_peaks_as_fasta(\n",
    "    bedtool=peaks_bedtool,\n",
    "    genome_fa=genome_fasta,\n",
    "    output_fasta=output_fasta\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of \"peaks\" found: 199\n"
     ]
    }
   ],
   "source": [
    "def get_positions(record, motif, slop):\n",
    "    \"\"\"\n",
    "    Iterates over a sequence and returns positions that match a motif.\n",
    "    \"\"\"\n",
    "    \n",
    "    positions = []\n",
    "    chrom, start, end, strand = record_id_to_coords(record.id)\n",
    "    i = 0\n",
    "    \n",
    "    while i < len(record.seq)-len(motif):\n",
    "        kmer = record.seq[i:i+len(motif)].upper()\n",
    "        if kmer == motif:\n",
    "            if strand == '+':\n",
    "                positions.append(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(\n",
    "                    chrom, i+start-slop, i+start+len(motif)+slop, motif, '0', strand\n",
    "                ))\n",
    "            elif strand == '-':\n",
    "                positions.append(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(\n",
    "                    chrom, end-i-len(motif)-slop, end-i+slop, motif, '0', strand\n",
    "                ))\n",
    "        i+=1\n",
    "    return positions\n",
    "\n",
    "\n",
    "positions = []\n",
    "records = SeqIO.parse(output_fasta, \"fasta\")\n",
    "for record in records:\n",
    "    positions += get_positions(record, motif, slop=10)\n",
    "\n",
    "print(\"Number of \\\"peaks\\\" found: {}\".format(len(positions)))\n",
    "with open(os.path.join(output_dir, os.path.splitext(os.path.basename(hist_genes))[0] + \".with.{}.bed\".format(motif)), 'w') as o:\n",
    "    for position in positions:\n",
    "        o.write(position)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make sure every sequence we have has the motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_sequence_records_with_motif(fn, motif):\n",
    "    \"\"\"\n",
    "    Returns a SeqIO record of sequences containing the motif. We expect this \n",
    "    FASTA file to contain the reverse-complement sequences if negative strand. \n",
    "    We also expect the FASTA headers to contain strand info \n",
    "    @see BedTool.sequence(fi=hg19.fa, s=True)\n",
    "    \"\"\"\n",
    "    seqs_with_motif = []\n",
    "    with open(fn, 'r') as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            chrom, start, end, strand = record_id_to_coords(record.id)\n",
    "            if (strand == '+' or strand =='-') and str(record.seq).upper().find(motif) != -1:\n",
    "                seqs_with_motif.append(record)\n",
    "            else:\n",
    "                assert strand == '+' or strand == '-'\n",
    "    return seqs_with_motif\n",
    "\n",
    "seqs_with_motif = return_sequence_records_with_motif(fn=os.path.join(output_dir, os.path.splitext(os.path.basename(hist_genes))[0] + \".with.{}.bed\".format(motif)), motif=motif)\n",
    "\n",
    "# Make sure every region describes a motif\n",
    "\n",
    "for record in seqs_with_motif:\n",
    "    try:\n",
    "        record_upper = record.seq.upper()\n",
    "        assert record_upper[record_upper.find(motif):record_upper.find(motif)+len(motif)] == motif\n",
    "    except AssertionError:\n",
    "        print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3-essential",
   "language": "python",
   "name": "python3-essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
