{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For this metagene plot, we're using either the top 200/1000 or Q1 genes from the boxplot figures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import gffutils\n",
    "import os\n",
    "import glob\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "conf = 0.5\n",
    "prefix = 'Q1'  # can be 'IRES', 'TORIN', 'Q1', or 'top200'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.bed',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.bed',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.bed']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_beds = sorted(glob.glob(os.path.join(input_dir, '*.conf{}.bed'.format(conf))))\n",
    "print(len(all_beds))\n",
    "all_beds[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 16275\n",
      "** Excluded (features not well defined or gene not found in gtf): 362\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 9489\n",
      "** Excluded (features not well defined or gene not found in gtf): 205\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 14402\n",
      "** Excluded (features not well defined or gene not found in gtf): 290\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 21127\n",
      "** Excluded (features not well defined or gene not found in gtf): 433\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 5230\n",
      "** Excluded (features not well defined or gene not found in gtf): 105\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 7311\n",
      "** Excluded (features not well defined or gene not found in gtf): 158\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 6218\n",
      "** Excluded (features not well defined or gene not found in gtf): 144\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 7811\n",
      "** Excluded (features not well defined or gene not found in gtf): 161\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ETF1-1000_S16_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/ETF1-1000_S16_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 25424\n",
      "** Excluded (features not well defined or gene not found in gtf): 504\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/GSTP1-1000_S17_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/GSTP1-1000_S17_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 39002\n",
      "** Excluded (features not well defined or gene not found in gtf): 730\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 149225\n",
      "** Excluded (features not well defined or gene not found in gtf): 2524\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 72410\n",
      "** Excluded (features not well defined or gene not found in gtf): 1189\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 51100\n",
      "** Excluded (features not well defined or gene not found in gtf): 852\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-T_1_S32_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-T_1_S32_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 66\n",
      "** Excluded (features not well defined or gene not found in gtf): 3\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-T_2_S33_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2-T_2_S33_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 21221\n",
      "** Excluded (features not well defined or gene not found in gtf): 398\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2_plusT_1_S34_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2_plusT_1_S34_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 38013\n",
      "** Excluded (features not well defined or gene not found in gtf): 666\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2_plusT_2_S35_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RBFOX2_plusT_2_S35_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 24045\n",
      "** Excluded (features not well defined or gene not found in gtf): 452\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 174739\n",
      "** Excluded (features not well defined or gene not found in gtf): 3054\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 228118\n",
      "** Excluded (features not well defined or gene not found in gtf): 4203\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 60702\n",
      "** Excluded (features not well defined or gene not found in gtf): 1204\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 41679\n",
      "** Excluded (features not well defined or gene not found in gtf): 829\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 44743\n",
      "** Excluded (features not well defined or gene not found in gtf): 935\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 22044\n",
      "** Excluded (features not well defined or gene not found in gtf): 500\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 20421\n",
      "** Excluded (features not well defined or gene not found in gtf): 486\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS3-1000_S19_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/RPS3-1000_S19_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 116486\n",
      "** Excluded (features not well defined or gene not found in gtf): 2333\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample13_S31_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample13_S31_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 3868\n",
      "** Excluded (features not well defined or gene not found in gtf): 73\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample14_S13_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample14_S13_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 9752\n",
      "** Excluded (features not well defined or gene not found in gtf): 210\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample19_S17_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample19_S17_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 37603\n",
      "** Excluded (features not well defined or gene not found in gtf): 637\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample20_S18_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample20_S18_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 30998\n",
      "** Excluded (features not well defined or gene not found in gtf): 521\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample29_S38_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample29_S38_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 6428\n",
      "** Excluded (features not well defined or gene not found in gtf): 137\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample30_S39_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample30_S39_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 4281\n",
      "** Excluded (features not well defined or gene not found in gtf): 84\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample31_S40_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample31_S40_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 8491\n",
      "** Excluded (features not well defined or gene not found in gtf): 182\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample32_S22_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample32_S22_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 12472\n",
      "** Excluded (features not well defined or gene not found in gtf): 239\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample41_S41_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample41_S41_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 12822\n",
      "** Excluded (features not well defined or gene not found in gtf): 340\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample42_S42_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample42_S42_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 25130\n",
      "** Excluded (features not well defined or gene not found in gtf): 519\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample5_S26_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample5_S26_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 5797\n",
      "** Excluded (features not well defined or gene not found in gtf): 113\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_33_S61_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_33_S61_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 17048\n",
      "** Excluded (features not well defined or gene not found in gtf): 346\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_34_S62_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_34_S62_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 20458\n",
      "** Excluded (features not well defined or gene not found in gtf): 402\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_35_S63_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_35_S63_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 14744\n",
      "** Excluded (features not well defined or gene not found in gtf): 283\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_36_S64_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_36_S64_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 12138\n",
      "** Excluded (features not well defined or gene not found in gtf): 227\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 19329\n",
      "** Excluded (features not well defined or gene not found in gtf): 397\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 26941\n",
      "** Excluded (features not well defined or gene not found in gtf): 549\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 28666\n",
      "** Excluded (features not well defined or gene not found in gtf): 582\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 20252\n",
      "** Excluded (features not well defined or gene not found in gtf): 435\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 34471\n",
      "** Excluded (features not well defined or gene not found in gtf): 709\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_39_S67_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_39_S67_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 13582\n",
      "** Excluded (features not well defined or gene not found in gtf): 308\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_40_S68_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_40_S68_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 13103\n",
      "** Excluded (features not well defined or gene not found in gtf): 278\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_6_S58_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Sample_6_S58_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 4205\n",
      "** Excluded (features not well defined or gene not found in gtf): 84\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 16521\n",
      "** Excluded (features not well defined or gene not found in gtf): 380\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 15424\n",
      "** Excluded (features not well defined or gene not found in gtf): 351\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 24989\n",
      "** Excluded (features not well defined or gene not found in gtf): 523\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 24829\n",
      "** Excluded (features not well defined or gene not found in gtf): 482\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Tia1_1000ng_1_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Tia1_1000ng_1_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 28139\n",
      "** Excluded (features not well defined or gene not found in gtf): 463\n",
      "------------------------------------------------------------------------------------------------------\n",
      "------------------------------------------------------------------------------------------------------\n",
      "module load bioperl;perl metaPlotR/rel_and_abs_dist_calc.pl --bed /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Tia1_1000ng_2_S27_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.annotated.Riboseq2.Q1.bed --regions /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/riboseq2.region_sizes.Q1.txt > /home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/Tia1_1000ng_2_S27_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\n",
      "** Total bedfile lines: 27682\n",
      "** Excluded (features not well defined or gene not found in gtf): 467\n",
      "------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "regions = os.path.join(input_dir, 'riboseq2.region_sizes.{}.txt'.format(prefix))\n",
    "for bed in all_beds:\n",
    "    # q1_bed = os.path.splitext(bed)[0] + \".annotated.GSE112353_Riboseq.{}.bed\".format(prefix)\n",
    "    q1_bed = os.path.splitext(bed)[0] + \".annotated.Riboseq2.{}.bed\".format(prefix)\n",
    "    # output_txt = os.path.join(output_dir, os.path.splitext(os.path.basename(bed))[0] + \".GSE112353_Riboseq.{}.dist.measures.txt\".format(prefix))\n",
    "    output_txt = os.path.join(output_dir, os.path.splitext(os.path.basename(bed))[0] + \".Riboseq2.{}.dist.measures.txt\".format(prefix))\n",
    "    cmd = 'module load bioperl;'\n",
    "    cmd += 'perl metaPlotR/rel_and_abs_dist_calc.pl --bed {} --regions {} > {}'.format(\n",
    "        q1_bed, \n",
    "        regions,\n",
    "        output_txt\n",
    "    )\n",
    "    print(\"------------------------------------------------------------------------------------------------------\")\n",
    "    print(cmd)\n",
    "    ! $cmd\n",
    "    print(\"------------------------------------------------------------------------------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3essential (tscc)",
   "language": "python",
   "name": "python3essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
