{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "library('scales')\n",
    "library('ggplot2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in and plot RPS2-1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd('/home/bay001/projects/kris_apobec_20200121/permanent_data/final_analysis/figure_3/inputs/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Q1, \"Deep\" datasets\n",
    "# rps2_dist_measures <- \"RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# apo_dist_measures <- \"ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# rbfox2_dist_measures <- \"RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "etf1_dist_measures <- \"ETF1-1000_S16_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "gstp1_dist_measures <- \"GSTP1-1000_S17_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "abce1_dist_measures <- \"ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "tia1_r1_dist_measures <- \"Tia1_1000ng_1_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "tia1_r2_dist_measures <- \"Tia1_1000ng_2_S27_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "\n",
    "# Q1, Samples datasets\n",
    "sample13_dist_measures <- \"Sample13_S31_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample29_dist_measures <- \"Sample29_S38_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample30_dist_measures <- \"Sample30_S39_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample31_dist_measures <- \"Sample31_S40_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample32_dist_measures <- \"Sample32_S22_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample33_dist_measures <- \"Sample_33_S61_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample34_dist_measures <- \"Sample_34_S62_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample35_dist_measures <- \"Sample_35_S63_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample36_dist_measures <- \"Sample_36_S64_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample37_dist_measures <- \"Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample38_dist_measures <- \"Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample39_dist_measures <- \"Sample_39_S67_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample40_dist_measures <- \"Sample_40_S68_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample41_dist_measures <- \"Sample41_S41_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample42_dist_measures <- \"Sample42_S42_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "\n",
    "# Q1, 10M downsampled.\n",
    "ti_55_10m_dist_measures <- \"TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "ti_56_dist_measures <- \"TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "ti_58_10m_dist_measures <- \"TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "ti_59_10m_dist_measures <- \"TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "\n",
    "# IRES, 10M downsampled.\n",
    "ti_55_10m_ires_dist_measures <- \"TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "ti_56_ires_dist_measures <- \"TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "sample37_10m_ires_dist_measures <- \"Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "sample38_10m_ires_dist_measures <- \"Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "\n",
    "# TORIN, 10M downsampled.\n",
    "ti_55_torin_10m_ires_dist_measures <- \"TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "ti_56_torin_dist_measures <- \"TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "ti_58_torin_10m_dist_measures <- \"TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "ti_59_torin_10m_dist_measures <- \"TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "sample37_torin_10m_ires_dist_measures <- \"Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "sample38_torin_10m_ires_dist_measures <- \"Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "\n",
    "# Q1, 21M downsampled.\n",
    "apo_21m_dist_measures <- \"ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sample37_21m_dist_measures <- \"Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "rps2_21m_dist_measures <- \"RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "\n",
    "# \"plusT\" samples\n",
    "# Apobec_plusT_1_dist_measures <- \"Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# Apobec_plusT_2_dist_measures <- \"Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# ApobecT_1_dist_measures <- \"Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# ApobecT_2_dist_measures <- \"Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# RPS2_plusT_1_dist_measures <- \"RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# RPS2_plusT_2_dist_measures <- \"RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# RPS2_T_1_dist_measures <- \"RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# RPS2_T_2_dist_measures <- \"RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "\n",
    "# \"plusT\" TORIN samples\n",
    "Apobec_plusT_1_torin_dist_measures <- \"Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "Apobec_plusT_2_torin_dist_measures <- \"Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "ApobecT_1_torin_dist_measures <- \"Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "ApobecT_2_torin_dist_measures <- \"Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "RPS2_plusT_1_torin_dist_measures <- \"RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "RPS2_plusT_2_torin_dist_measures <- \"RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "RPS2_T_1_torin_dist_measures <- \"RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "RPS2_T_2_torin_dist_measures <- \"RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
    "\n",
    "# \"plusT\" IRES samples\n",
    "Apobec_plusT_1_ires_dist_measures <- \"Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "Apobec_plusT_2_ires_dist_measures <- \"Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "ApobecT_1_ires_dist_measures <- \"Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "ApobecT_2_ires_dist_measures <- \"Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "RPS2_plusT_1_ires_dist_measures <- \"RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "RPS2_plusT_2_ires_dist_measures <- \"RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "RPS2_T_1_ires_dist_measures <- \"RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "RPS2_T_2_ires_dist_measures <- \"RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
    "\n",
    "# Single cell samples:\n",
    "# sc1_dist_measures <- \"RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# sc2_dist_measures <- \"RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# sc3_dist_measures <- \"APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "# sc4_dist_measures <- \"RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
    "sc1_dist_measures <- \"RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "sc2_dist_measures <- \"RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "sc3_dist_measures <- \"APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "sc4_dist_measures <- \"RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "\n",
    "# Q1, \"Deep\" datasets (Riboseq2)\n",
    "rps2_dist_measures <- \"RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "apo_dist_measures <- \"ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "rbfox2_dist_measures <- \"RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "\n",
    "# \"plusT\" samples (Riboseq2)\n",
    "Apobec_plusT_1_dist_measures <- \"Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "Apobec_plusT_2_dist_measures <- \"Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "ApobecT_1_dist_measures <- \"Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "ApobecT_2_dist_measures <- \"Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "RPS2_plusT_1_dist_measures <- \"RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "RPS2_plusT_2_dist_measures <- \"RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "RPS2_T_1_dist_measures <- \"RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "RPS2_T_2_dist_measures <- \"RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
    "\n",
    "\n",
    "all_dist_measures <- c(\n",
    "    sc1_dist_measures,\n",
    "    sc2_dist_measures,\n",
    "    sc3_dist_measures,\n",
    "    sc4_dist_measures,\n",
    "    Apobec_plusT_1_ires_dist_measures,\n",
    "    Apobec_plusT_2_ires_dist_measures,\n",
    "    ApobecT_1_ires_dist_measures,\n",
    "    ApobecT_2_ires_dist_measures,\n",
    "    RPS2_plusT_1_ires_dist_measures,\n",
    "    RPS2_plusT_2_ires_dist_measures,\n",
    "    RPS2_T_1_ires_dist_measures,\n",
    "    RPS2_T_2_ires_dist_measures,\n",
    "    Apobec_plusT_1_torin_dist_measures,\n",
    "    Apobec_plusT_2_torin_dist_measures,\n",
    "    ApobecT_1_torin_dist_measures,\n",
    "    ApobecT_2_torin_dist_measures,\n",
    "    RPS2_plusT_1_torin_dist_measures,\n",
    "    RPS2_plusT_2_torin_dist_measures,\n",
    "    RPS2_T_1_torin_dist_measures,\n",
    "    RPS2_T_2_torin_dist_measures,\n",
    "    Apobec_plusT_1_dist_measures,\n",
    "    Apobec_plusT_2_dist_measures,\n",
    "    ApobecT_1_dist_measures,\n",
    "    ApobecT_2_dist_measures,\n",
    "    RPS2_plusT_1_dist_measures,\n",
    "    RPS2_plusT_2_dist_measures,\n",
    "    RPS2_T_1_dist_measures,\n",
    "    RPS2_T_2_dist_measures,\n",
    "    ti_55_torin_10m_ires_dist_measures,\n",
    "    ti_56_torin_dist_measures,\n",
    "    ti_58_torin_10m_dist_measures,\n",
    "    ti_59_torin_10m_dist_measures,\n",
    "    sample37_torin_10m_ires_dist_measures,\n",
    "    sample38_torin_10m_ires_dist_measures,\n",
    "    rps2_dist_measures,\n",
    "    apo_dist_measures,\n",
    "    rbfox2_dist_measures,\n",
    "    etf1_dist_measures,\n",
    "    gstp1_dist_measures,\n",
    "    abce1_dist_measures,\n",
    "    tia1_r1_dist_measures,\n",
    "    tia1_r2_dist_measures,\n",
    "    sample13_dist_measures,\n",
    "    sample29_dist_measures,\n",
    "    sample30_dist_measures,\n",
    "    sample31_dist_measures,\n",
    "    sample32_dist_measures,\n",
    "    sample33_dist_measures,\n",
    "    sample34_dist_measures,\n",
    "    sample35_dist_measures,\n",
    "    sample36_dist_measures,\n",
    "    sample37_dist_measures,\n",
    "    sample38_dist_measures,\n",
    "    sample39_dist_measures,\n",
    "    sample40_dist_measures,\n",
    "    sample41_dist_measures,\n",
    "    sample42_dist_measures,\n",
    "    ti_55_10m_dist_measures,\n",
    "    ti_56_dist_measures,\n",
    "    ti_58_10m_dist_measures,\n",
    "    ti_59_10m_dist_measures,\n",
    "    ti_55_10m_ires_dist_measures,\n",
    "    ti_56_ires_dist_measures,\n",
    "    sample37_10m_ires_dist_measures,\n",
    "    sample38_10m_ires_dist_measures,\n",
    "    apo_21m_dist_measures,\n",
    "    sample37_21m_dist_measures,\n",
    "    rps2_21m_dist_measures\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"DIST PRE TX DEDUP FOR RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 71221    14\n",
      "[1] \"DIST POST TX DEDUP FOR RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-6_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 71221    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 50248    14\n",
      "[1] \"DIST POST TX DEDUP FOR RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD-7_barcodes.tsv.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 50248    14\n",
      "[1] \"DIST PRE TX DEDUP FOR APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 9284   14\n",
      "[1] \"DIST POST TX DEDUP FOR APOBEC-STAMP_possorted_genome_bam-APOBEC_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 9284   14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 59498    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-STAMP_possorted_genome_bam-RPS2_STAMP_Apo_filtered_lenti_common_expression_barcodes.txt.fx.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 59498    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 30 14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 30 14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 58 14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 58 14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 26 14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 26 14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 43 14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 43 14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 111  14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 111  14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 121  14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 121  14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 221  14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 221  14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 247  14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 247  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 155  14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 155  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 141  14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 141  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 98 14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 98 14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 128  14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 128  14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1047   14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1047   14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1027   14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1027   14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1172   14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1172   14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1476   14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 1476   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 6074   14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_1_S30_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 6074   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 7650   14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec_plusT_2_S31_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 7650   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 5125   14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_1_S28_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 5125   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 7153   14\n",
      "[1] \"DIST POST TX DEDUP FOR Apobec-T_2_S29_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 7153   14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 21544    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_1_S38_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 21544    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 19935    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2_plusT_2_S39_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 19935    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 40850    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_1_S36_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 40850    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 43808    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-T_2_S37_L003_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 43808    14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 825  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 825  14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 732  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 732  14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 766  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 766  14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 768  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 768  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 730  14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 730  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 719  14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.TORIN.dist.measures.txt\"\n",
      "[1] 719  14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 223915     14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 223915     14\n",
      "[1] \"DIST PRE TX DEDUP FOR ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 20694    14\n",
      "[1] \"DIST POST TX DEDUP FOR ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 20694    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 146701     14\n",
      "[1] \"DIST POST TX DEDUP FOR RBFOX2-1000_S20_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.Riboseq2.Q1.dist.measures.txt\"\n",
      "[1] 146701     14\n",
      "[1] \"DIST PRE TX DEDUP FOR ETF1-1000_S16_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14730    14\n",
      "[1] \"DIST POST TX DEDUP FOR ETF1-1000_S16_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14725    14\n",
      "[1] \"DIST PRE TX DEDUP FOR GSTP1-1000_S17_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 22641    14\n",
      "[1] \"DIST POST TX DEDUP FOR GSTP1-1000_S17_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 22631    14\n",
      "[1] \"DIST PRE TX DEDUP FOR ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 10175    14\n",
      "[1] \"DIST POST TX DEDUP FOR ABCE1-1000_S15_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 10171    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Tia1_1000ng_1_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14625    14\n",
      "[1] \"DIST POST TX DEDUP FOR Tia1_1000ng_1_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14624    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Tia1_1000ng_2_S27_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14502    14\n",
      "[1] \"DIST POST TX DEDUP FOR Tia1_1000ng_2_S27_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14497    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample13_S31_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 2410   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample13_S31_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 2408   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample29_S38_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 3982   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample29_S38_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 3977   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample30_S39_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 2642   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample30_S39_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 2640   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample31_S40_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 4822   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample31_S40_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 4822   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample32_S22_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7332   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample32_S22_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7329   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_33_S61_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9814   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_33_S61_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9810   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_34_S62_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 11970    14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_34_S62_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 11966    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_35_S63_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9385   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_35_S63_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9380   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_36_S64_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7676   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_36_S64_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7672   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 17060    14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 17052    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 20786    14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 20775    14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_39_S67_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8568   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_39_S67_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8563   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_40_S68_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8364   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_40_S68_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8362   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample41_S41_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7565   14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample41_S41_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 7563   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample42_S42_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14401    14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample42_S42_merged.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14398    14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9356   14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 9355   14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8951   14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8948   14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14697    14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_58_S15_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14687    14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14451    14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_59_S16_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 14445    14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 100  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_55_S13_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 100  14\n",
      "[1] \"DIST PRE TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 117  14\n",
      "[1] \"DIST POST TX DEDUP FOR TI_56_S14_L001_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 117  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 102  14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 102  14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 114  14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_38_S66_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.10M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.IRES.dist.measures.txt\"\n",
      "[1] 114  14\n",
      "[1] \"DIST PRE TX DEDUP FOR ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8702   14\n",
      "[1] \"DIST POST TX DEDUP FOR ApoControl-1000_S21_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 8700   14\n",
      "[1] \"DIST PRE TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 15983    14\n",
      "[1] \"DIST POST TX DEDUP FOR Sample_37_S65_L006_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 15975    14\n",
      "[1] \"DIST PRE TX DEDUP FOR RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 97899    14\n",
      "[1] \"DIST POST TX DEDUP FOR RPS2-1000_S18_L002_R1_001.fastqTr.sorted.STARUnmapped.out.sorted.STARAligned.out.sorted.21M_a0_b0_e0.01.conf0.5.GSE112353_Riboseq.Q1.dist.measures.txt\"\n",
      "[1] 97856    14\n"
     ]
    }
   ],
   "source": [
    "for (dist_measures in all_dist_measures) {\n",
    "    dist <- read.delim (dist_measures, header = T)\n",
    "\n",
    "    dim(dist)\n",
    "    print(paste0(\"DIST PRE TX DEDUP FOR \", dist_measures))\n",
    "    print(dim(dist))\n",
    "    # Determine longest length transcript for each gene.\n",
    "    \n",
    "    # I guess we don't really need this since we're already selecting one transcript per gene in notebook 02.\n",
    "    # However, notebook 02 assumes hg38-gene annotations, and keeping this in will de-dup transcripts based on \n",
    "    # hg19-gene annotations, which we will be carrying forward. This results in a very minor reduction in our \n",
    "    # edit list (ie. both ENST00000409145.1 and ENST00000410042.1 belong to MRPL30 in hg19 but refer to \n",
    "    # different genes MRPL30-005 and RP11-38C17.1-001 in hg38). Not a big difference either way, but it's \n",
    "    # worth noting since the numbers don't 100% match up. \n",
    "    \n",
    "    trx_len <- dist$utr5_size + dist$cds_size + dist$utr3_size # Determine transcript length\n",
    "    temp <- data.frame(dist$gene_name, dist$refseqID, trx_len)\n",
    "    colnames(temp) <- c(\"gene_name\", \"gid\", \"trx_len\") \n",
    "    temp <- temp[order(temp$gene_name, -temp$trx_len),]\n",
    "    temp <- temp[!duplicated(temp$gene_name),]\n",
    "    # limit m6a data to one transcript per gene (longest)\n",
    "    dist <- dist[dist$refseqID %in% temp$gid,]\n",
    "    print(paste0(\"DIST POST TX DEDUP FOR \", dist_measures))\n",
    "    print(dim(dist))\n",
    "    # View size of our dataset (rows, columns)\n",
    "    dim(dist)\n",
    "\n",
    "    qplot(dist$rel_location, geom=\"histogram\") + geom_vline(xintercept = 1:2, col = \"grey\") + theme_bw()\n",
    "    summary(data.frame(dist$utr5_size, dist$cds_size, dist$utr3_size))\n",
    "\n",
    "    utr5.SF <- median(dist$utr5_size, na.rm = T)/median(dist$cds_size, na.rm = T)\n",
    "    utr3.SF <- median(dist$utr3_size, na.rm = T)/median(dist$cds_size, na.rm = T)\n",
    "\n",
    "    # assign the regions to new dataframes\n",
    "    utr5.dist <- dist[dist$rel_location < 1, ]\n",
    "    cds.dist <- dist [dist$rel_location < 2 & dist$rel_location >= 1, ]\n",
    "    utr3.dist <- dist[dist$rel_location >= 2, ]\n",
    "\n",
    "    utr5.dist$rel_location <- rescale(utr5.dist$rel_location, to = c(1-utr5.SF, 1), from = c(0,1))\n",
    "    utr3.dist$rel_location <- rescale(utr3.dist$rel_location, to = c(2, 2+utr3.SF), from = c(2,3))\n",
    "\n",
    "    # Combine and plot\n",
    "    ## Histogram\n",
    "    metagene.coord <- c(utr5.dist$rel_location, cds.dist$rel_location, utr3.dist$rel_location)\n",
    "    p <- qplot(metagene.coord, geom=\"histogram\") + geom_vline(xintercept = 1:2, col = \"grey\") + theme_bw()\n",
    "\n",
    "    qplot(metagene.coord, geom=\"freqpoly\") + geom_vline(xintercept = 1:2, col = \"red\") + theme_bw()\n",
    "\n",
    "    df <- data.frame(metagene.coord, 'red')\n",
    "    ggplot(df) + geom_density(aes(x = metagene.coord, colour = 'red')) + xlim(0, 3) + \n",
    "      theme_bw() + geom_vline(xintercept = 1:2, col = \"grey\")\n",
    "\n",
    "    write.table(dist$rel_location, paste0(dist_measures, '.rel_location.txt'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"dim\"  \"6074\" \"14\"  \n",
      "[1] \"dim\"  \"6074\" \"14\"  \n",
      "[1] \"dim\"  \"7650\" \"14\"  \n",
      "[1] \"dim\"  \"7650\" \"14\"  \n",
      "[1] \"dim\"  \"5125\" \"14\"  \n",
      "[1] \"dim\"  \"5125\" \"14\"  \n",
      "[1] \"dim\"  \"7153\" \"14\"  \n",
      "[1] \"dim\"  \"7153\" \"14\"  \n",
      "[1] \"dim\"   \"21544\" \"14\"   \n",
      "[1] \"dim\"   \"21544\" \"14\"   \n",
      "[1] \"dim\"   \"19935\" \"14\"   \n",
      "[1] \"dim\"   \"19935\" \"14\"   \n",
      "[1] \"dim\"   \"40850\" \"14\"   \n",
      "[1] \"dim\"   \"40850\" \"14\"   \n",
      "[1] \"dim\"   \"43808\" \"14\"   \n",
      "[1] \"dim\"   \"43808\" \"14\"   \n"
     ]
    }
   ],
   "source": [
    "all_torin_measures <- c(\n",
    "    Apobec_plusT_1_dist_measures,\n",
    "    Apobec_plusT_2_dist_measures,\n",
    "    ApobecT_1_dist_measures,\n",
    "    ApobecT_2_dist_measures,\n",
    "    RPS2_plusT_1_dist_measures,\n",
    "    RPS2_plusT_2_dist_measures,\n",
    "    RPS2_T_1_dist_measures,\n",
    "    RPS2_T_2_dist_measures\n",
    ")\n",
    "\n",
    "metagene.list <- vector(\"list\", length(all_torin_measures))\n",
    "i <- 1\n",
    "\n",
    "for (dist_measures in all_torin_measures) {\n",
    "    dist <- read.delim (dist_measures, header = T)\n",
    "\n",
    "    print(c('dim', dim(dist)))\n",
    "\n",
    "    ### Determine longest length transcript for each gene. ###\n",
    "\n",
    "    trx_len <- dist$utr5_size + dist$cds_size + dist$utr3_size # Determine transcript length\n",
    "    temp <- data.frame(dist$gene_name, dist$refseqID, trx_len)\n",
    "    colnames(temp) <- c(\"gene_name\", \"gid\", \"trx_len\") \n",
    "    temp <- temp[order(temp$gene_name, -temp$trx_len),]\n",
    "    temp\n",
    "    temp <- temp[!duplicated(temp$gene_name),]\n",
    "    temp\n",
    "    # limit m6a data to one transcript per gene (longest)\n",
    "    dist <- dist[dist$refseqID %in% temp$gid,]\n",
    "\n",
    "    # View size of our dataset (rows, columns). Shouldn't change since we remove overlapping/duplicate entries beforehand.\n",
    "    print(c('dim', dim(dist)))\n",
    "    qplot(dist$rel_location, geom=\"histogram\") + geom_vline(xintercept = 1:2, col = \"grey\") + theme_bw()\n",
    "    summary(data.frame(dist$utr5_size, dist$cds_size, dist$utr3_size))\n",
    "\n",
    "    utr5.SF <- median(dist$utr5_size, na.rm = T)/median(dist$cds_size, na.rm = T)\n",
    "    utr3.SF <- median(dist$utr3_size, na.rm = T)/median(dist$cds_size, na.rm = T)\n",
    "\n",
    "    # assign the regions to new dataframes\n",
    "    utr5.dist <- dist[dist$rel_location < 1, ]\n",
    "    cds.dist <- dist [dist$rel_location < 2 & dist$rel_location >= 1, ]\n",
    "    utr3.dist <- dist[dist$rel_location >= 2, ]\n",
    "\n",
    "    utr5.dist$rel_location <- rescale(utr5.dist$rel_location, to = c(1-utr5.SF, 1), from = c(0,1))\n",
    "    utr3.dist$rel_location <- rescale(utr3.dist$rel_location, to = c(2, 2+utr3.SF), from = c(2,3))\n",
    "\n",
    "    # Combine and plot\n",
    "    ## Histogram\n",
    "    metagene.coord <- c(utr5.dist$rel_location, cds.dist$rel_location, utr3.dist$rel_location)\n",
    "    metagene.list[[i]] <- metagene.coord\n",
    "    i <- i + 1\n",
    "    # p <- qplot(metagene.coord, geom=\"histogram\") + geom_vline(xintercept = 1:2, col = \"grey\") + theme_bw()\n",
    "\n",
    "    # qplot(metagene.coord, geom=\"freqpoly\") + geom_vline(xintercept = 1:2, col = \"red\") + theme_bw()\n",
    "\n",
    "    # df <- data.frame(metagene.coord, 'red')\n",
    "    # ggplot(df) + geom_density(aes(x = metagene.coord, colour = 'red')) + xlim(0, 3) + \n",
    "    #   theme_bw() + geom_vline(xintercept = 1:2, col = \"grey\")\n",
    "\n",
    "    # write.table(dist$rel_location, paste0(dist_measures, '.rel_location.txt'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metagene.cord <- c(\n",
    "    metagene.list[[1]], metagene.list[[2]], metagene.list[[3]], metagene.list[[4]], \n",
    "    metagene.list[[5]], metagene.list[[6]], metagene.list[[7]], metagene.list[[8]]\n",
    ")\n",
    "mod <- c(rep(\"Apobec_plusT_1\", length(metagene.list[[1]])), \n",
    "         rep(\"Apobec_plusT_2\", length(metagene.list[[2]])),\n",
    "         rep(\"ApobecT_1\", length(metagene.list[[3]])),\n",
    "         rep(\"ApobecT_2\", length(metagene.list[[4]])),\n",
    "         rep(\"RPS2_plusT_1\", length(metagene.list[[5]])), \n",
    "         rep(\"RPS2_plusT_2\", length(metagene.list[[6]])),\n",
    "         rep(\"RPS2T_1\", length(metagene.list[[7]])),\n",
    "         rep(\"RPS2T_2\", length(metagene.list[[8]]))\n",
    "        )\n",
    "df <- data.frame(metagene.cord, mod)\n",
    "png('test.png')\n",
    "ggplot(df) + geom_density(aes(x = metagene.cord, colour = mod)) + xlim(0, 3) + \n",
    "  theme_bw() + geom_vline(xintercept = 1:2, col = \"grey\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metagene.cord <- c(\n",
    "    metagene.list[[5]], metagene.list[[6]], metagene.list[[7]], metagene.list[[8]]\n",
    ")\n",
    "mod <- c(\n",
    "    rep(\"RPS2_plusT_1\", length(metagene.list[[5]])), \n",
    "    rep(\"RPS2_plusT_2\", length(metagene.list[[6]])),\n",
    "    rep(\"RPS2T_1\", length(metagene.list[[7]])),\n",
    "    rep(\"RPS2T_2\", length(metagene.list[[8]]))\n",
    ")\n",
    "df <- data.frame(metagene.cord, mod)\n",
    "png('testRps.png')\n",
    "ggplot(df) + geom_density(aes(x = metagene.cord, colour = mod)) + xlim(0, 3) + \n",
    "  theme_bw() + geom_vline(xintercept = 1:2, col = \"grey\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metagene.cord <- c(\n",
    "    metagene.list[[1]], metagene.list[[2]], metagene.list[[3]], metagene.list[[4]]\n",
    ")\n",
    "mod <- c(\n",
    "    rep(\"Apobec_plusT_1\", length(metagene.list[[1]])), \n",
    "    rep(\"Apobec_plusT_2\", length(metagene.list[[2]])),\n",
    "    rep(\"ApobecT_1\", length(metagene.list[[3]])),\n",
    "    rep(\"ApobecT_2\", length(metagene.list[[4]]))\n",
    ")\n",
    "df <- data.frame(metagene.cord, mod)\n",
    "png('testApo.png')\n",
    "ggplot(df) + geom_density(aes(x = metagene.cord, colour = mod)) + xlim(0, 3) + \n",
    "  theme_bw() + geom_vline(xintercept = 1:2, col = \"grey\")\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R-Seurat-fc4a4f5",
   "language": "R",
   "name": "r_seurat_fc4a4f5"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
