{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook parses 10X input (possorted_genome.bam WITH MD TAGs) and randomly downsamples to a specified number of barcodes.\n",
    "- functionally equivalent to all of the other 01_ notebooks, but we're just subsetting on different groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import pysam\n",
    "import time\n",
    "from collections import defaultdict\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "from qtools import Submitter\n",
    "\n",
    "pd.set_option('display.max_colwidth', 1000)\n",
    "pd.set_option('display.width', 10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_file = os.path.join(input_dir, '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD.bam')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in expression values from Alex to help us decide which barcodes to look for/split.\n",
    "\n",
    "original sources: \n",
    "```\n",
    "RPS2_STAMP_Apo_filtered_lenti_common_expression.csv\n",
    "APOBEC_STAMP_Apo_filtered_lenti_common_expression.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tia_barcodes = pd.read_csv('/home/iachaim/notebooks/10X/APOBEC/02_02_Subsetting_on_non_APO_overlap/results/For_Motif_TIA1_no_APO_edits_barcodes.csv')\n",
    "rbfox2_barcodes = pd.read_csv('/home/iachaim/notebooks/10X/APOBEC/02_02_Subsetting_on_non_APO_overlap/results/For_Motif_RBFOX2_no_APO_edits_barcodes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>batch</th>\n",
       "      <th>new_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACAGGAAGTC-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAGTATGTCCA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAAAGGCTGAAC-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAACACAGACGA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAACACATACTG-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        index            batch new_clusters\n",
       "0  possorted_genome_bam_MD-AAACCCACAGGAAGTC-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP\n",
       "1  possorted_genome_bam_MD-AAACCCAGTATGTCCA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP\n",
       "2  possorted_genome_bam_MD-AAACGAAAGGCTGAAC-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP\n",
       "3  possorted_genome_bam_MD-AAACGAACACAGACGA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP\n",
       "4  possorted_genome_bam_MD-AAACGAACACATACTG-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tia_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>batch</th>\n",
       "      <th>new_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAAGAGCCCAA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAAGTGCTACT-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACATCTATCT-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACATCTCAAG-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAGTAGATGTA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        index            batch  new_clusters\n",
       "0  possorted_genome_bam_MD-AAACCCAAGAGCCCAA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP\n",
       "1  possorted_genome_bam_MD-AAACCCAAGTGCTACT-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP\n",
       "2  possorted_genome_bam_MD-AAACCCACATCTATCT-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP\n",
       "3  possorted_genome_bam_MD-AAACCCACATCTCAAG-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP\n",
       "4  possorted_genome_bam_MD-AAACCCAGTAGATGTA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rbfox2_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def index_to_barcode(row, to_replace):\n",
    "    \"\"\"\n",
    "    Not really a true index, just a column named 'index'.\n",
    "    \"\"\"\n",
    "    barcode = row['index']\n",
    "    for replace in to_replace:\n",
    "        barcode = barcode.replace(replace,'')\n",
    "        \n",
    "    ### Just to make sure we're grabbing and correctly formatting barcodes.\n",
    "    assert barcode.endswith('-1')\n",
    "\n",
    "    return barcode\n",
    "\n",
    "tia_barcodes['barcode'] = tia_barcodes.apply(index_to_barcode, args=(['possorted_genome_bam_MD-', '-RBFOX_TIA-Edits'],), axis=1)\n",
    "rbfox2_barcodes['barcode'] = rbfox2_barcodes.apply(index_to_barcode, args=(['possorted_genome_bam_MD-', '-RBFOX_TIA-Edits'],), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>batch</th>\n",
       "      <th>new_clusters</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACAGGAAGTC-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "      <td>AAACCCACAGGAAGTC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAGTATGTCCA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "      <td>AAACCCAGTATGTCCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAAAGGCTGAAC-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "      <td>AAACGAAAGGCTGAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAACACAGACGA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "      <td>AAACGAACACAGACGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_MD-AAACGAACACATACTG-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>TIA1_STAMP</td>\n",
       "      <td>AAACGAACACATACTG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        index            batch new_clusters             barcode\n",
       "0  possorted_genome_bam_MD-AAACCCACAGGAAGTC-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP  AAACCCACAGGAAGTC-1\n",
       "1  possorted_genome_bam_MD-AAACCCAGTATGTCCA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP  AAACCCAGTATGTCCA-1\n",
       "2  possorted_genome_bam_MD-AAACGAAAGGCTGAAC-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP  AAACGAAAGGCTGAAC-1\n",
       "3  possorted_genome_bam_MD-AAACGAACACAGACGA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP  AAACGAACACAGACGA-1\n",
       "4  possorted_genome_bam_MD-AAACGAACACATACTG-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits   TIA1_STAMP  AAACGAACACATACTG-1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tia_barcodes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>batch</th>\n",
       "      <th>new_clusters</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAAGAGCCCAA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "      <td>AAACCCAAGAGCCCAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAAGTGCTACT-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "      <td>AAACCCAAGTGCTACT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACATCTATCT-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "      <td>AAACCCACATCTATCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCACATCTCAAG-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "      <td>AAACCCACATCTCAAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_MD-AAACCCAGTAGATGTA-1-RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX_TIA-Edits</td>\n",
       "      <td>RBFOX2_STAMP</td>\n",
       "      <td>AAACCCAGTAGATGTA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        index            batch  new_clusters             barcode\n",
       "0  possorted_genome_bam_MD-AAACCCAAGAGCCCAA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP  AAACCCAAGAGCCCAA-1\n",
       "1  possorted_genome_bam_MD-AAACCCAAGTGCTACT-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP  AAACCCAAGTGCTACT-1\n",
       "2  possorted_genome_bam_MD-AAACCCACATCTATCT-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP  AAACCCACATCTATCT-1\n",
       "3  possorted_genome_bam_MD-AAACCCACATCTCAAG-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP  AAACCCACATCTCAAG-1\n",
       "4  possorted_genome_bam_MD-AAACCCAGTAGATGTA-1-RBFOX_TIA-Edits  RBFOX_TIA-Edits  RBFOX2_STAMP  AAACCCAGTAGATGTA-1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rbfox2_barcodes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downsample RBFOX2 barcodes from 844 down to 1 and run motif analysis. Let's see how far down we can go and still recover the RBFOX2 motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These come from 01_group_RBFOX2_TIA_barcodes.ipynb\n",
    "tia_barcodes_file = os.path.join(output_dir, '33_barcodes.tsv')\n",
    "rbfox2_barcodes_file = os.path.join(output_dir, '34_barcodes.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "50bf3575709d459184075ce260ae5561",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=590), HTML(value=u'')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rand 0 trial 0 exists. Skipping.\n",
      "rand 0 trial 1 exists. Skipping.\n",
      "rand 0 trial 2 exists. Skipping.\n",
      "rand 0 trial 3 exists. Skipping.\n",
      "rand 0 trial 4 exists. Skipping.\n",
      "rand 0 trial 5 exists. Skipping.\n",
      "rand 0 trial 6 exists. Skipping.\n",
      "rand 0 trial 7 exists. Skipping.\n",
      "rand 0 trial 8 exists. Skipping.\n",
      "rand 0 trial 9 exists. Skipping.\n",
      "rand 1 trial 0 exists. Skipping.\n",
      "rand 1 trial 1 exists. Skipping.\n",
      "rand 1 trial 2 exists. Skipping.\n",
      "rand 1 trial 3 exists. Skipping.\n",
      "rand 1 trial 4 exists. Skipping.\n",
      "rand 1 trial 5 exists. Skipping.\n",
      "rand 1 trial 6 exists. Skipping.\n",
      "rand 1 trial 7 exists. Skipping.\n",
      "rand 1 trial 8 exists. Skipping.\n",
      "rand 1 trial 9 exists. Skipping.\n",
      "rand 2 trial 0 exists. Skipping.\n",
      "rand 2 trial 1 exists. Skipping.\n",
      "rand 2 trial 2 exists. Skipping.\n",
      "rand 2 trial 3 exists. Skipping.\n",
      "rand 2 trial 4 exists. Skipping.\n",
      "rand 2 trial 5 exists. Skipping.\n",
      "rand 2 trial 6 exists. Skipping.\n",
      "rand 2 trial 7 exists. Skipping.\n",
      "rand 2 trial 8 exists. Skipping.\n",
      "rand 2 trial 9 exists. Skipping.\n",
      "rand 3 trial 0 exists. Skipping.\n",
      "rand 3 trial 1 exists. Skipping.\n",
      "rand 3 trial 2 exists. Skipping.\n",
      "rand 3 trial 3 exists. Skipping.\n",
      "rand 3 trial 4 exists. Skipping.\n",
      "rand 3 trial 5 exists. Skipping.\n",
      "rand 3 trial 6 exists. Skipping.\n",
      "rand 3 trial 7 exists. Skipping.\n",
      "rand 3 trial 8 exists. Skipping.\n",
      "rand 3 trial 9 exists. Skipping.\n",
      "rand 4 trial 0 exists. Skipping.\n",
      "rand 4 trial 1 exists. Skipping.\n",
      "rand 4 trial 2 exists. Skipping.\n",
      "rand 4 trial 3 exists. Skipping.\n",
      "rand 4 trial 4 exists. Skipping.\n",
      "rand 4 trial 5 exists. Skipping.\n",
      "rand 4 trial 6 exists. Skipping.\n",
      "rand 4 trial 7 exists. Skipping.\n",
      "rand 4 trial 8 exists. Skipping.\n",
      "rand 4 trial 9 exists. Skipping.\n",
      "rand 5 trial 0 exists. Skipping.\n",
      "rand 5 trial 1 exists. Skipping.\n",
      "rand 5 trial 2 exists. Skipping.\n",
      "rand 5 trial 3 exists. Skipping.\n",
      "rand 5 trial 4 exists. Skipping.\n",
      "rand 5 trial 5 exists. Skipping.\n",
      "rand 5 trial 6 exists. Skipping.\n",
      "rand 5 trial 7 exists. Skipping.\n",
      "rand 5 trial 8 exists. Skipping.\n",
      "rand 5 trial 9 exists. Skipping.\n",
      "rand 6 trial 0 exists. Skipping.\n",
      "rand 6 trial 1 exists. Skipping.\n",
      "rand 6 trial 2 exists. Skipping.\n",
      "rand 6 trial 3 exists. Skipping.\n",
      "rand 6 trial 4 exists. Skipping.\n",
      "rand 6 trial 5 exists. Skipping.\n",
      "rand 6 trial 6 exists. Skipping.\n",
      "rand 6 trial 7 exists. Skipping.\n",
      "rand 6 trial 8 exists. Skipping.\n",
      "rand 6 trial 9 exists. Skipping.\n",
      "rand 7 trial 0 exists. Skipping.\n",
      "rand 7 trial 1 exists. Skipping.\n",
      "rand 7 trial 2 exists. Skipping.\n",
      "rand 7 trial 3 exists. Skipping.\n",
      "rand 7 trial 4 exists. Skipping.\n",
      "rand 7 trial 5 exists. Skipping.\n",
      "rand 7 trial 6 exists. Skipping.\n",
      "rand 7 trial 7 exists. Skipping.\n",
      "rand 7 trial 8 exists. Skipping.\n",
      "rand 7 trial 9 exists. Skipping.\n",
      "rand 8 trial 0 exists. Skipping.\n",
      "rand 8 trial 1 exists. Skipping.\n",
      "rand 8 trial 2 exists. Skipping.\n",
      "rand 8 trial 3 exists. Skipping.\n",
      "rand 8 trial 4 exists. Skipping.\n",
      "rand 8 trial 5 exists. Skipping.\n",
      "rand 8 trial 6 exists. Skipping.\n",
      "rand 8 trial 7 exists. Skipping.\n",
      "rand 8 trial 8 exists. Skipping.\n",
      "rand 8 trial 9 exists. Skipping.\n",
      "rand 9 trial 0 exists. Skipping.\n",
      "rand 9 trial 1 exists. Skipping.\n",
      "rand 9 trial 2 exists. Skipping.\n",
      "rand 9 trial 3 exists. Skipping.\n",
      "rand 9 trial 4 exists. Skipping.\n",
      "rand 9 trial 5 exists. Skipping.\n",
      "rand 9 trial 6 exists. Skipping.\n",
      "rand 9 trial 7 exists. Skipping.\n",
      "rand 9 trial 8 exists. Skipping.\n",
      "rand 9 trial 9 exists. Skipping.\n",
      "rand 10 trial 0 exists. Skipping.\n",
      "rand 10 trial 1 exists. Skipping.\n",
      "rand 10 trial 2 exists. Skipping.\n",
      "rand 10 trial 3 exists. Skipping.\n",
      "rand 10 trial 4 exists. Skipping.\n",
      "rand 10 trial 5 exists. Skipping.\n",
      "rand 10 trial 6 exists. Skipping.\n",
      "rand 10 trial 7 exists. Skipping.\n",
      "rand 10 trial 8 exists. Skipping.\n",
      "rand 10 trial 9 exists. Skipping.\n",
      "rand 11 trial 0 exists. Skipping.\n",
      "rand 11 trial 1 exists. Skipping.\n",
      "rand 11 trial 2 exists. Skipping.\n",
      "rand 11 trial 3 exists. Skipping.\n",
      "rand 11 trial 4 exists. Skipping.\n",
      "rand 11 trial 5 exists. Skipping.\n",
      "rand 11 trial 6 exists. Skipping.\n",
      "rand 11 trial 7 exists. Skipping.\n",
      "rand 11 trial 8 exists. Skipping.\n",
      "rand 11 trial 9 exists. Skipping.\n",
      "rand 12 trial 0 exists. Skipping.\n",
      "rand 12 trial 1 exists. Skipping.\n",
      "rand 12 trial 2 exists. Skipping.\n",
      "rand 12 trial 3 exists. Skipping.\n",
      "rand 12 trial 4 exists. Skipping.\n",
      "rand 12 trial 5 exists. Skipping.\n",
      "rand 12 trial 6 exists. Skipping.\n",
      "rand 12 trial 7 exists. Skipping.\n",
      "rand 12 trial 8 exists. Skipping.\n",
      "rand 12 trial 9 exists. Skipping.\n",
      "rand 13 trial 0 exists. Skipping.\n",
      "rand 13 trial 1 exists. Skipping.\n",
      "rand 13 trial 2 exists. Skipping.\n",
      "rand 13 trial 3 exists. Skipping.\n",
      "rand 13 trial 4 exists. Skipping.\n",
      "rand 13 trial 5 exists. Skipping.\n",
      "rand 13 trial 6 exists. Skipping.\n",
      "rand 13 trial 7 exists. Skipping.\n",
      "rand 13 trial 8 exists. Skipping.\n",
      "rand 13 trial 9 exists. Skipping.\n",
      "rand 14 trial 0 exists. Skipping.\n",
      "rand 14 trial 1 exists. Skipping.\n",
      "rand 14 trial 2 exists. Skipping.\n",
      "rand 14 trial 3 exists. Skipping.\n",
      "rand 14 trial 4 exists. Skipping.\n",
      "rand 14 trial 5 exists. Skipping.\n",
      "rand 14 trial 6 exists. Skipping.\n",
      "rand 14 trial 7 exists. Skipping.\n",
      "rand 14 trial 8 exists. Skipping.\n",
      "rand 14 trial 9 exists. Skipping.\n",
      "rand 15 trial 0 exists. Skipping.\n",
      "rand 15 trial 1 exists. Skipping.\n",
      "rand 15 trial 2 exists. Skipping.\n",
      "rand 15 trial 3 exists. Skipping.\n",
      "rand 15 trial 4 exists. Skipping.\n",
      "rand 15 trial 5 exists. Skipping.\n",
      "rand 15 trial 6 exists. Skipping.\n",
      "rand 15 trial 7 exists. Skipping.\n",
      "rand 15 trial 8 exists. Skipping.\n",
      "rand 15 trial 9 exists. Skipping.\n",
      "rand 16 trial 0 exists. Skipping.\n",
      "rand 16 trial 1 exists. Skipping.\n",
      "rand 16 trial 2 exists. Skipping.\n",
      "rand 16 trial 3 exists. Skipping.\n",
      "rand 16 trial 4 exists. Skipping.\n",
      "rand 16 trial 5 exists. Skipping.\n",
      "rand 16 trial 6 exists. Skipping.\n",
      "rand 16 trial 7 exists. Skipping.\n",
      "rand 16 trial 8 exists. Skipping.\n",
      "rand 16 trial 9 exists. Skipping.\n",
      "rand 17 trial 0 exists. Skipping.\n",
      "rand 17 trial 1 exists. Skipping.\n",
      "rand 17 trial 2 exists. Skipping.\n",
      "rand 17 trial 3 exists. Skipping.\n",
      "rand 17 trial 4 exists. Skipping.\n",
      "rand 17 trial 5 exists. Skipping.\n",
      "rand 17 trial 6 exists. Skipping.\n",
      "rand 17 trial 7 exists. Skipping.\n",
      "rand 17 trial 8 exists. Skipping.\n",
      "rand 17 trial 9 exists. Skipping.\n",
      "rand 18 trial 0 exists. Skipping.\n",
      "rand 18 trial 1 exists. Skipping.\n",
      "rand 18 trial 2 exists. Skipping.\n",
      "rand 18 trial 3 exists. Skipping.\n",
      "rand 18 trial 4 exists. Skipping.\n",
      "rand 18 trial 5 exists. Skipping.\n",
      "rand 18 trial 6 exists. Skipping.\n",
      "rand 18 trial 7 exists. Skipping.\n",
      "rand 18 trial 8 exists. Skipping.\n",
      "rand 18 trial 9 exists. Skipping.\n",
      "rand 19 trial 0 exists. Skipping.\n",
      "rand 19 trial 1 exists. Skipping.\n",
      "rand 19 trial 2 exists. Skipping.\n",
      "rand 19 trial 3 exists. Skipping.\n",
      "rand 19 trial 4 exists. Skipping.\n",
      "rand 19 trial 5 exists. Skipping.\n",
      "rand 19 trial 6 exists. Skipping.\n",
      "rand 19 trial 7 exists. Skipping.\n",
      "rand 19 trial 8 exists. Skipping.\n",
      "rand 19 trial 9 exists. Skipping.\n",
      "rand 20 trial 0 exists. Skipping.\n",
      "rand 20 trial 1 exists. Skipping.\n",
      "rand 20 trial 2 exists. Skipping.\n",
      "rand 20 trial 3 exists. Skipping.\n",
      "rand 20 trial 4 exists. Skipping.\n",
      "rand 20 trial 5 exists. Skipping.\n",
      "rand 20 trial 6 exists. Skipping.\n",
      "rand 20 trial 7 exists. Skipping.\n",
      "rand 20 trial 8 exists. Skipping.\n",
      "rand 20 trial 9 exists. Skipping.\n",
      "rand 21 trial 0 exists. Skipping.\n",
      "rand 21 trial 1 exists. Skipping.\n",
      "rand 21 trial 2 exists. Skipping.\n",
      "rand 21 trial 3 exists. Skipping.\n",
      "rand 21 trial 4 exists. Skipping.\n",
      "rand 21 trial 5 exists. Skipping.\n",
      "rand 21 trial 6 exists. Skipping.\n",
      "rand 21 trial 7 exists. Skipping.\n",
      "rand 21 trial 8 exists. Skipping.\n",
      "rand 21 trial 9 exists. Skipping.\n",
      "rand 22 trial 0 exists. Skipping.\n",
      "rand 22 trial 1 exists. Skipping.\n",
      "rand 22 trial 2 exists. Skipping.\n",
      "rand 22 trial 3 exists. Skipping.\n",
      "rand 22 trial 4 exists. Skipping.\n",
      "rand 22 trial 5 exists. Skipping.\n",
      "rand 22 trial 6 exists. Skipping.\n",
      "rand 22 trial 7 exists. Skipping.\n",
      "rand 22 trial 8 exists. Skipping.\n",
      "rand 22 trial 9 exists. Skipping.\n",
      "rand 23 trial 0 exists. Skipping.\n",
      "rand 23 trial 1 exists. Skipping.\n",
      "rand 23 trial 2 exists. Skipping.\n",
      "rand 23 trial 3 exists. Skipping.\n",
      "rand 23 trial 4 exists. Skipping.\n",
      "rand 23 trial 5 exists. Skipping.\n",
      "rand 23 trial 6 exists. Skipping.\n",
      "rand 23 trial 7 exists. Skipping.\n",
      "rand 23 trial 8 exists. Skipping.\n",
      "rand 23 trial 9 exists. Skipping.\n",
      "rand 24 trial 0 exists. Skipping.\n",
      "rand 24 trial 1 exists. Skipping.\n",
      "rand 24 trial 2 exists. Skipping.\n",
      "rand 24 trial 3 exists. Skipping.\n",
      "rand 24 trial 4 exists. Skipping.\n",
      "rand 24 trial 5 exists. Skipping.\n",
      "rand 24 trial 6 exists. Skipping.\n",
      "rand 24 trial 7 exists. Skipping.\n",
      "rand 24 trial 8 exists. Skipping.\n",
      "rand 24 trial 9 exists. Skipping.\n",
      "rand 25 trial 0 exists. Skipping.\n",
      "rand 25 trial 1 exists. Skipping.\n",
      "rand 25 trial 2 exists. Skipping.\n",
      "rand 25 trial 3 exists. Skipping.\n",
      "rand 25 trial 4 exists. Skipping.\n",
      "rand 25 trial 5 exists. Skipping.\n",
      "rand 25 trial 6 exists. Skipping.\n",
      "rand 25 trial 7 exists. Skipping.\n",
      "rand 25 trial 8 exists. Skipping.\n",
      "rand 25 trial 9 exists. Skipping.\n",
      "rand 26 trial 0 exists. Skipping.\n",
      "rand 26 trial 1 exists. Skipping.\n",
      "rand 26 trial 2 exists. Skipping.\n",
      "rand 26 trial 3 exists. Skipping.\n",
      "rand 26 trial 4 exists. Skipping.\n",
      "rand 26 trial 5 exists. Skipping.\n",
      "rand 26 trial 6 exists. Skipping.\n",
      "rand 26 trial 7 exists. Skipping.\n",
      "rand 26 trial 8 exists. Skipping.\n",
      "rand 26 trial 9 exists. Skipping.\n",
      "rand 27 trial 0 exists. Skipping.\n",
      "rand 27 trial 1 exists. Skipping.\n",
      "rand 27 trial 2 exists. Skipping.\n",
      "rand 27 trial 3 exists. Skipping.\n",
      "rand 27 trial 4 exists. Skipping.\n",
      "rand 27 trial 5 exists. Skipping.\n",
      "rand 27 trial 6 exists. Skipping.\n",
      "rand 27 trial 7 exists. Skipping.\n",
      "rand 27 trial 8 exists. Skipping.\n",
      "rand 27 trial 9 exists. Skipping.\n",
      "rand 28 trial 0 exists. Skipping.\n",
      "rand 28 trial 1 exists. Skipping.\n",
      "rand 28 trial 2 exists. Skipping.\n",
      "rand 28 trial 3 exists. Skipping.\n",
      "rand 28 trial 4 exists. Skipping.\n",
      "rand 28 trial 5 exists. Skipping.\n",
      "rand 28 trial 6 exists. Skipping.\n",
      "rand 28 trial 7 exists. Skipping.\n",
      "rand 28 trial 8 exists. Skipping.\n",
      "rand 28 trial 9 exists. Skipping.\n",
      "rand 29 trial 0 exists. Skipping.\n",
      "rand 29 trial 1 exists. Skipping.\n",
      "rand 29 trial 2 exists. Skipping.\n",
      "rand 29 trial 3 exists. Skipping.\n",
      "rand 29 trial 4 exists. Skipping.\n",
      "rand 29 trial 5 exists. Skipping.\n",
      "rand 29 trial 6 exists. Skipping.\n",
      "rand 29 trial 7 exists. Skipping.\n",
      "rand 29 trial 8 exists. Skipping.\n",
      "rand 29 trial 9 exists. Skipping.\n",
      "rand 30 trial 0 exists. Skipping.\n",
      "rand 30 trial 1 exists. Skipping.\n",
      "rand 30 trial 2 exists. Skipping.\n",
      "rand 30 trial 3 exists. Skipping.\n",
      "rand 30 trial 4 exists. Skipping.\n",
      "rand 30 trial 5 exists. Skipping.\n",
      "rand 30 trial 6 exists. Skipping.\n",
      "rand 30 trial 7 exists. Skipping.\n",
      "rand 30 trial 8 exists. Skipping.\n",
      "rand 30 trial 9 exists. Skipping.\n",
      "rand 31 trial 0 exists. Skipping.\n",
      "rand 31 trial 1 exists. Skipping.\n",
      "rand 31 trial 2 exists. Skipping.\n",
      "rand 31 trial 3 exists. Skipping.\n",
      "rand 31 trial 4 exists. Skipping.\n",
      "rand 31 trial 5 exists. Skipping.\n",
      "rand 31 trial 6 exists. Skipping.\n",
      "rand 31 trial 7 exists. Skipping.\n",
      "rand 31 trial 8 exists. Skipping.\n",
      "rand 31 trial 9 exists. Skipping.\n",
      "rand 32 trial 0 exists. Skipping.\n",
      "rand 32 trial 1 exists. Skipping.\n",
      "rand 32 trial 2 exists. Skipping.\n",
      "rand 32 trial 3 exists. Skipping.\n",
      "rand 32 trial 4 exists. Skipping.\n",
      "rand 32 trial 5 exists. Skipping.\n",
      "rand 32 trial 6 exists. Skipping.\n",
      "rand 32 trial 7 exists. Skipping.\n",
      "rand 32 trial 8 exists. Skipping.\n",
      "rand 32 trial 9 exists. Skipping.\n",
      "rand 33 trial 0 exists. Skipping.\n",
      "rand 33 trial 1 exists. Skipping.\n",
      "rand 33 trial 2 exists. Skipping.\n",
      "rand 33 trial 3 exists. Skipping.\n",
      "rand 33 trial 4 exists. Skipping.\n",
      "rand 33 trial 5 exists. Skipping.\n",
      "rand 33 trial 6 exists. Skipping.\n",
      "rand 33 trial 7 exists. Skipping.\n",
      "rand 33 trial 8 exists. Skipping.\n",
      "rand 33 trial 9 exists. Skipping.\n",
      "rand 34 trial 0 exists. Skipping.\n",
      "rand 34 trial 1 exists. Skipping.\n",
      "rand 34 trial 2 exists. Skipping.\n",
      "rand 34 trial 3 exists. Skipping.\n",
      "rand 34 trial 4 exists. Skipping.\n",
      "rand 34 trial 5 exists. Skipping.\n",
      "rand 34 trial 6 exists. Skipping.\n",
      "rand 34 trial 7 exists. Skipping.\n",
      "rand 34 trial 8 exists. Skipping.\n",
      "rand 34 trial 9 exists. Skipping.\n",
      "rand 35 trial 0 exists. Skipping.\n",
      "rand 35 trial 1 exists. Skipping.\n",
      "rand 35 trial 2 exists. Skipping.\n",
      "rand 35 trial 3 exists. Skipping.\n",
      "rand 35 trial 4 exists. Skipping.\n",
      "rand 35 trial 5 exists. Skipping.\n",
      "rand 35 trial 6 exists. Skipping.\n",
      "rand 35 trial 7 exists. Skipping.\n",
      "rand 35 trial 8 exists. Skipping.\n",
      "rand 35 trial 9 exists. Skipping.\n",
      "rand 36 trial 0 exists. Skipping.\n",
      "rand 36 trial 1 exists. Skipping.\n",
      "rand 36 trial 2 exists. Skipping.\n",
      "rand 36 trial 3 exists. Skipping.\n",
      "rand 36 trial 4 exists. Skipping.\n",
      "rand 36 trial 5 exists. Skipping.\n",
      "rand 36 trial 6 exists. Skipping.\n",
      "rand 36 trial 7 exists. Skipping.\n",
      "rand 36 trial 8 exists. Skipping.\n",
      "rand 36 trial 9 exists. Skipping.\n",
      "rand 37 trial 0 exists. Skipping.\n",
      "rand 37 trial 1 exists. Skipping.\n",
      "rand 37 trial 2 exists. Skipping.\n",
      "rand 37 trial 3 exists. Skipping.\n",
      "rand 37 trial 4 exists. Skipping.\n",
      "rand 37 trial 5 exists. Skipping.\n",
      "rand 37 trial 6 exists. Skipping.\n",
      "rand 37 trial 7 exists. Skipping.\n",
      "rand 37 trial 8 exists. Skipping.\n",
      "rand 37 trial 9 exists. Skipping.\n",
      "rand 38 trial 0 exists. Skipping.\n",
      "rand 38 trial 1 exists. Skipping.\n",
      "rand 38 trial 2 exists. Skipping.\n",
      "rand 38 trial 3 exists. Skipping.\n",
      "rand 38 trial 4 exists. Skipping.\n",
      "rand 38 trial 5 exists. Skipping.\n",
      "rand 38 trial 6 exists. Skipping.\n",
      "rand 38 trial 7 exists. Skipping.\n",
      "rand 38 trial 8 exists. Skipping.\n",
      "rand 38 trial 9 exists. Skipping.\n",
      "rand 39 trial 0 exists. Skipping.\n",
      "rand 39 trial 1 exists. Skipping.\n",
      "rand 39 trial 2 exists. Skipping.\n",
      "rand 39 trial 3 exists. Skipping.\n",
      "rand 39 trial 4 exists. Skipping.\n",
      "rand 39 trial 5 exists. Skipping.\n",
      "rand 39 trial 6 exists. Skipping.\n",
      "rand 39 trial 7 exists. Skipping.\n",
      "rand 39 trial 8 exists. Skipping.\n",
      "rand 39 trial 9 exists. Skipping.\n",
      "rand 40 trial 0 exists. Skipping.\n",
      "rand 40 trial 1 exists. Skipping.\n",
      "rand 40 trial 2 exists. Skipping.\n",
      "rand 40 trial 3 exists. Skipping.\n",
      "rand 40 trial 4 exists. Skipping.\n",
      "rand 40 trial 5 exists. Skipping.\n",
      "rand 40 trial 6 exists. Skipping.\n",
      "rand 40 trial 7 exists. Skipping.\n",
      "rand 40 trial 8 exists. Skipping.\n",
      "rand 40 trial 9 exists. Skipping.\n",
      "rand 41 trial 0 exists. Skipping.\n",
      "rand 41 trial 1 exists. Skipping.\n",
      "rand 41 trial 2 exists. Skipping.\n",
      "rand 41 trial 3 exists. Skipping.\n",
      "rand 41 trial 4 exists. Skipping.\n",
      "rand 41 trial 5 exists. Skipping.\n",
      "rand 41 trial 6 exists. Skipping.\n",
      "rand 41 trial 7 exists. Skipping.\n",
      "rand 41 trial 8 exists. Skipping.\n",
      "rand 41 trial 9 exists. Skipping.\n",
      "rand 42 trial 0 exists. Skipping.\n",
      "rand 42 trial 1 exists. Skipping.\n",
      "rand 42 trial 2 exists. Skipping.\n",
      "rand 42 trial 3 exists. Skipping.\n",
      "rand 42 trial 4 exists. Skipping.\n",
      "rand 42 trial 5 exists. Skipping.\n",
      "rand 42 trial 6 exists. Skipping.\n",
      "rand 42 trial 7 exists. Skipping.\n",
      "rand 42 trial 8 exists. Skipping.\n",
      "rand 42 trial 9 exists. Skipping.\n",
      "rand 43 trial 0 exists. Skipping.\n",
      "rand 43 trial 1 exists. Skipping.\n",
      "rand 43 trial 2 exists. Skipping.\n",
      "rand 43 trial 3 exists. Skipping.\n",
      "rand 43 trial 4 exists. Skipping.\n",
      "rand 43 trial 5 exists. Skipping.\n",
      "rand 43 trial 6 exists. Skipping.\n",
      "rand 43 trial 7 exists. Skipping.\n",
      "rand 43 trial 8 exists. Skipping.\n",
      "rand 43 trial 9 exists. Skipping.\n",
      "rand 44 trial 0 exists. Skipping.\n",
      "rand 44 trial 1 exists. Skipping.\n",
      "rand 44 trial 2 exists. Skipping.\n",
      "rand 44 trial 3 exists. Skipping.\n",
      "rand 44 trial 4 exists. Skipping.\n",
      "rand 44 trial 5 exists. Skipping.\n",
      "rand 44 trial 6 exists. Skipping.\n",
      "rand 44 trial 7 exists. Skipping.\n",
      "rand 44 trial 8 exists. Skipping.\n",
      "rand 44 trial 9 exists. Skipping.\n",
      "rand 45 trial 0 exists. Skipping.\n",
      "rand 45 trial 1 exists. Skipping.\n",
      "rand 45 trial 2 exists. Skipping.\n",
      "rand 45 trial 3 exists. Skipping.\n",
      "rand 45 trial 4 exists. Skipping.\n",
      "rand 45 trial 5 exists. Skipping.\n",
      "rand 45 trial 6 exists. Skipping.\n",
      "rand 45 trial 7 exists. Skipping.\n",
      "rand 45 trial 8 exists. Skipping.\n",
      "rand 45 trial 9 exists. Skipping.\n",
      "rand 46 trial 0 exists. Skipping.\n",
      "rand 46 trial 1 exists. Skipping.\n",
      "rand 46 trial 2 exists. Skipping.\n",
      "rand 46 trial 3 exists. Skipping.\n",
      "rand 46 trial 4 exists. Skipping.\n",
      "rand 46 trial 5 exists. Skipping.\n",
      "rand 46 trial 6 exists. Skipping.\n",
      "rand 46 trial 7 exists. Skipping.\n",
      "rand 46 trial 8 exists. Skipping.\n",
      "rand 46 trial 9 exists. Skipping.\n",
      "rand 47 trial 0 exists. Skipping.\n",
      "rand 47 trial 1 exists. Skipping.\n",
      "rand 47 trial 2 exists. Skipping.\n",
      "rand 47 trial 3 exists. Skipping.\n",
      "rand 47 trial 4 exists. Skipping.\n",
      "rand 47 trial 5 exists. Skipping.\n",
      "rand 47 trial 6 exists. Skipping.\n",
      "rand 47 trial 7 exists. Skipping.\n",
      "rand 47 trial 8 exists. Skipping.\n",
      "rand 47 trial 9 exists. Skipping.\n",
      "rand 48 trial 0 exists. Skipping.\n",
      "rand 48 trial 1 exists. Skipping.\n",
      "rand 48 trial 2 exists. Skipping.\n",
      "rand 48 trial 3 exists. Skipping.\n",
      "rand 48 trial 4 exists. Skipping.\n",
      "rand 48 trial 5 exists. Skipping.\n",
      "rand 48 trial 6 exists. Skipping.\n",
      "rand 48 trial 7 exists. Skipping.\n",
      "rand 48 trial 8 exists. Skipping.\n",
      "rand 48 trial 9 exists. Skipping.\n",
      "rand 49 trial 0 exists. Skipping.\n",
      "rand 49 trial 1 exists. Skipping.\n",
      "rand 49 trial 2 exists. Skipping.\n",
      "rand 49 trial 3 exists. Skipping.\n",
      "rand 49 trial 4 exists. Skipping.\n",
      "rand 49 trial 5 exists. Skipping.\n",
      "rand 49 trial 6 exists. Skipping.\n",
      "rand 49 trial 7 exists. Skipping.\n",
      "rand 49 trial 8 exists. Skipping.\n",
      "rand 49 trial 9 exists. Skipping.\n",
      "rand 844 trial 0 exists. Skipping.\n",
      "rand 844 trial 1 exists. Skipping.\n",
      "rand 844 trial 2 exists. Skipping.\n",
      "rand 844 trial 3 exists. Skipping.\n",
      "rand 844 trial 4 exists. Skipping.\n",
      "rand 844 trial 5 exists. Skipping.\n",
      "rand 844 trial 6 exists. Skipping.\n",
      "rand 844 trial 7 exists. Skipping.\n",
      "rand 844 trial 8 exists. Skipping.\n",
      "rand 844 trial 9 exists. Skipping.\n",
      "rand 800 trial 0 exists. Skipping.\n",
      "rand 800 trial 1 exists. Skipping.\n",
      "rand 800 trial 2 exists. Skipping.\n",
      "rand 800 trial 3 exists. Skipping.\n",
      "rand 800 trial 4 exists. Skipping.\n",
      "rand 800 trial 5 exists. Skipping.\n",
      "rand 800 trial 6 exists. Skipping.\n",
      "rand 800 trial 7 exists. Skipping.\n",
      "rand 800 trial 8 exists. Skipping.\n",
      "rand 800 trial 9 exists. Skipping.\n",
      "rand 700 trial 0 exists. Skipping.\n",
      "rand 700 trial 1 exists. Skipping.\n",
      "rand 700 trial 2 exists. Skipping.\n",
      "rand 700 trial 3 exists. Skipping.\n",
      "rand 700 trial 4 exists. Skipping.\n",
      "rand 700 trial 5 exists. Skipping.\n",
      "rand 700 trial 6 exists. Skipping.\n",
      "rand 700 trial 7 exists. Skipping.\n",
      "rand 700 trial 8 exists. Skipping.\n",
      "rand 700 trial 9 exists. Skipping.\n",
      "rand 600 trial 0 exists. Skipping.\n",
      "rand 600 trial 1 exists. Skipping.\n",
      "rand 600 trial 2 exists. Skipping.\n",
      "rand 600 trial 3 exists. Skipping.\n",
      "rand 600 trial 4 exists. Skipping.\n",
      "rand 600 trial 5 exists. Skipping.\n",
      "rand 600 trial 6 exists. Skipping.\n",
      "rand 600 trial 7 exists. Skipping.\n",
      "rand 600 trial 8 exists. Skipping.\n",
      "rand 600 trial 9 exists. Skipping.\n",
      "rand 500 trial 0 exists. Skipping.\n",
      "rand 500 trial 1 exists. Skipping.\n",
      "rand 500 trial 2 exists. Skipping.\n",
      "rand 500 trial 3 exists. Skipping.\n",
      "rand 500 trial 4 exists. Skipping.\n",
      "rand 500 trial 5 exists. Skipping.\n",
      "rand 500 trial 6 exists. Skipping.\n",
      "rand 500 trial 7 exists. Skipping.\n",
      "rand 500 trial 8 exists. Skipping.\n",
      "rand 500 trial 9 exists. Skipping.\n",
      "rand 400 trial 0 exists. Skipping.\n",
      "rand 400 trial 1 exists. Skipping.\n",
      "rand 400 trial 2 exists. Skipping.\n",
      "rand 400 trial 3 exists. Skipping.\n",
      "rand 400 trial 4 exists. Skipping.\n",
      "rand 400 trial 5 exists. Skipping.\n",
      "rand 400 trial 6 exists. Skipping.\n",
      "rand 400 trial 7 exists. Skipping.\n",
      "rand 400 trial 8 exists. Skipping.\n",
      "rand 400 trial 9 exists. Skipping.\n",
      "rand 300 trial 0 exists. Skipping.\n",
      "rand 300 trial 1 exists. Skipping.\n",
      "rand 300 trial 2 exists. Skipping.\n",
      "rand 300 trial 3 exists. Skipping.\n",
      "rand 300 trial 4 exists. Skipping.\n",
      "rand 300 trial 5 exists. Skipping.\n",
      "rand 300 trial 6 exists. Skipping.\n",
      "rand 300 trial 7 exists. Skipping.\n",
      "rand 300 trial 8 exists. Skipping.\n",
      "rand 300 trial 9 exists. Skipping.\n",
      "rand 200 trial 0 exists. Skipping.\n",
      "rand 200 trial 1 exists. Skipping.\n",
      "rand 200 trial 2 exists. Skipping.\n",
      "rand 200 trial 3 exists. Skipping.\n",
      "rand 200 trial 4 exists. Skipping.\n",
      "rand 200 trial 5 exists. Skipping.\n",
      "rand 200 trial 6 exists. Skipping.\n",
      "rand 200 trial 7 exists. Skipping.\n",
      "rand 200 trial 8 exists. Skipping.\n",
      "rand 200 trial 9 exists. Skipping.\n",
      "rand 100 trial 0 exists. Skipping.\n",
      "rand 100 trial 1 exists. Skipping.\n",
      "rand 100 trial 2 exists. Skipping.\n",
      "rand 100 trial 3 exists. Skipping.\n",
      "rand 100 trial 4 exists. Skipping.\n",
      "rand 100 trial 5 exists. Skipping.\n",
      "rand 100 trial 6 exists. Skipping.\n",
      "rand 100 trial 7 exists. Skipping.\n",
      "rand 100 trial 8 exists. Skipping.\n",
      "rand 100 trial 9 exists. Skipping.\n",
      "We have 590 files created.\n"
     ]
    }
   ],
   "source": [
    "ncells = [range(50) + [844, 800, 700, 600, 500, 400, 300, 200, 100]]\n",
    "# ncells = [range(50) + [844, 800, 700, 600, 500, 400, 300, 200, 100], range(50) + [518, 500, 400, 300, 200, 100]]  # in case we want to do this for TIA1 too.\n",
    "\n",
    "# barcodes_files = [rbfox2_barcodes_file, tia_barcodes_file]  # in case we want to do this for TIA1 too.\n",
    "barcodes_files = [rbfox2_barcodes_file]\n",
    "\n",
    "\n",
    "trials = range(10)\n",
    "counter = 0\n",
    "for barcodes_file, ncell in zip(barcodes_files, ncells):\n",
    "    progress = tnrange(len(ncell)*len(trials))\n",
    "    df = pd.read_csv(barcodes_file, names=['barcode'])\n",
    "    for n in ncell:\n",
    "        for t in trials:\n",
    "            if not os.path.exists(os.path.splitext(barcodes_file)[0] + \".rand{}.trial-{}.txt\".format(n, t)):  # We don't want to re-randomize files that already exist, harder to debug if they're constantly being overwritten.\n",
    "                dfs = df.sample(n=n) \n",
    "                dfs.to_csv(os.path.splitext(barcodes_file)[0] + \".rand{}.trial-{}.txt\".format(n, t), header=False, index=False)\n",
    "            else:\n",
    "                print(\"rand {} trial {} exists. Skipping.\".format(n, t))\n",
    "            counter += 1\n",
    "            progress.update(1)\n",
    "print(\"We have {} files created.\".format(counter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subtract capture sequence barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(540, 867)\n"
     ]
    }
   ],
   "source": [
    "tia_cs_file = '/home/iachaim/notebooks/10X/APOBEC/02_02_Subsetting_on_non_APO_overlap/results/TIA1_ORF_matrix.csv'\n",
    "tia_cs = pd.read_csv(tia_cs_file, index_col=0)\n",
    "tia_cs = set(tia_cs[tia_cs['TIA1_ORF'] > 0].index)\n",
    "\n",
    "rbfox2_cs_file = '/home/iachaim/notebooks/10X/APOBEC/02_02_Subsetting_on_non_APO_overlap/results/RBFOX2_ORF_matrix.csv'\n",
    "rbfox2_cs = pd.read_csv(rbfox2_cs_file, index_col=0)\n",
    "rbfox2_cs = set(rbfox2_cs[rbfox2_cs['RBFOX2_ORF'] > 0].index)\n",
    "\n",
    "print(len(tia_cs), len(rbfox2_cs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2089, 4627)\n"
     ]
    }
   ],
   "source": [
    "tia_barcodes = tia_barcodes[tia_barcodes['barcode'].isin(tia_cs)==False]\n",
    "rbfox2_barcodes = rbfox2_barcodes[rbfox2_barcodes['barcode'].isin(rbfox2_cs)==False]\n",
    "\n",
    "print(tia_barcodes.shape[0], rbfox2_barcodes.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "tia_barcodes_file = os.path.join(output_dir, '6_barcodes_noCS.txt')\n",
    "tia_barcodes[['barcode']].to_csv(tia_barcodes_file, sep='\\t', index=False, header=False)\n",
    "\n",
    "rbfox2_barcodes_file = os.path.join(output_dir, '7_barcodes_noCS.txt')\n",
    "rbfox2_barcodes[['barcode']].to_csv(rbfox2_barcodes_file, sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downsample (noCS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand10.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand50.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand100.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand200.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand300.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.rand500.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand10.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand50.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand100.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand200.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand300.txt, already made.\n",
      "skipping /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.rand500.txt, already made.\n"
     ]
    }
   ],
   "source": [
    "ncells = [10, 50, 100, 200, 300, 500]\n",
    "for barcodes_file in [tia_barcodes_file, rbfox2_barcodes_file]:\n",
    "    df = pd.read_csv(barcodes_file, names=['barcode'])\n",
    "    for n in ncells:\n",
    "        if not os.path.exists(os.path.splitext(barcodes_file)[0] + \".rand{}.txt\".format(n)):\n",
    "            dfs = df.sample(n=n) \n",
    "            dfs.to_csv(os.path.splitext(barcodes_file)[0] + \".rand{}.txt\".format(n), header=False, index=False)\n",
    "        else:\n",
    "            print(\"skipping {}, already made.\".format(os.path.splitext(barcodes_file)[0] + \".rand{}.txt\".format(n)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collect all the barcode lists from above and generate the split bam commands for each one. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "590\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/34_barcodes.rand0.trial-0.txt',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/34_barcodes.rand0.trial-1.txt',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/34_barcodes.rand0.trial-2.txt',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/34_barcodes.rand0.trial-3.txt',\n",
       " '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/34_barcodes.rand0.trial-4.txt']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_rand_barcode_groups = sorted(glob.glob(os.path.join(output_dir, '*.rand*trial*.txt')))\n",
    "print(len(all_rand_barcode_groups))\n",
    "all_rand_barcode_groups[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_commandline_for_grouping(bam_file, barcodes_file, output_dir, v=False):\n",
    "    cmd = 'module load python3essential;/home/bay001/projects/kris_apobec_20200121/scripts/split_bams.py '\n",
    "    cmd += '--possorted_bam_file {} '.format(bam_file)\n",
    "    cmd += '--barcodes_file {} '.format(barcodes_file)\n",
    "    cmd += '--output_dir {} '.format(output_dir)\n",
    "    cmd += '--group'\n",
    "    # --v means to SUBTRACT barcodes from a BAM file. \n",
    "    if v:\n",
    "        cmd += ' --v'\n",
    "    return cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmds = []\n",
    "for group in all_rand_barcode_groups:\n",
    "    if not os.path.exists(os.path.join(output_dir, os.path.splitext(os.path.basename(bam_file))[0] + '-{}.bam'.format(os.path.basename(group)))):\n",
    "        cmds.append(generate_commandline_for_grouping(bam_file, group, output_dir))\n",
    "len(cmds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Writing 0 tasks as an array-job.\n",
      "Wrote commands to subset_barcodes.sh.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<qtools.submitter.Submitter at 0x2b72c8971b10>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Submitter(cmds, 'subset_barcodes', array=True, nodes=1, ppn=1, walltime='24:00:00', submit=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset BAM file (minus CS barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['module load python3essential;/home/bay001/projects/kris_apobec_20200121/scripts/split_bams.py --possorted_bam_file /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD.bam --barcodes_file /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/6_barcodes_noCS.txt --output_dir /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/ --group',\n",
       " 'module load python3essential;/home/bay001/projects/kris_apobec_20200121/scripts/split_bams.py --possorted_bam_file /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD.bam --barcodes_file /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/7_barcodes_noCS.txt --output_dir /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/bam_file_groups/ --group']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmds = []\n",
    "for barcodes_file in [tia_barcodes_file, rbfox2_barcodes_file]:\n",
    "    cmds.append(generate_commandline_for_grouping(bam_file, barcodes_file, output_dir))\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Writing 2 tasks as an array-job.\n",
      "Wrote commands to subset_barcodes2.sh.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<qtools.submitter.Submitter at 0x2b72c8971110>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Submitter(cmds, 'subset_barcodes2', array=True, nodes=1, ppn=1, walltime='8:00:00', submit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python2-essential",
   "language": "python",
   "name": "python2-essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
