{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploratory notebook that removes C>T edited reads from a BAM file. Alex will check the expression profiles with and without C>T edited reads and see if UMAP differences are derived from edits or expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam \n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_softclip(cigar):\n",
    "    \"\"\"\n",
    "    Returns the number of bases to be softclipped on either left or right\n",
    "    side. Or both. If not softclipped, return 0\n",
    "    :param cigar: string\n",
    "        BAM/SAM CIGAR string\n",
    "    :return left: int\n",
    "        number of softclipped reads at the beginning\n",
    "    :return right: int\n",
    "        number of softclipped reads at the end\n",
    "    \"\"\"\n",
    "    softclip_regex = ur\"(\\d+)S\"\n",
    "    softclip = re.findall(softclip_regex,cigar)\n",
    "\n",
    "    softclip_right_regex = ur\"[\\w\\d]{1}(\\d+)S\" # if the softclip comes from the right side\n",
    "    softclip_right = re.findall(softclip_right_regex,cigar)\n",
    "\n",
    "    left = 0\n",
    "    right = 0\n",
    "\n",
    "    if softclip:\n",
    "        if len(softclip) == 2: # softclipped on both sides\n",
    "            left = int(softclip[0])\n",
    "            right = int(softclip[1])\n",
    "        elif len(softclip_right) == 1: # softclipped only on the RIGHT side\n",
    "            right = int(softclip[0])\n",
    "        else: # softclipped only on the LEFT side\n",
    "            left = int(softclip[0])\n",
    "    return left, right\n",
    "\n",
    "\n",
    "def remove_softclipped_reads(left, right, read_seq):\n",
    "    \"\"\"\n",
    "    Returns the read after removing softclipped bases.\n",
    "    :param left: int\n",
    "        left softclip\n",
    "    :param right: int\n",
    "        right softclip\n",
    "    :param read_seq: string\n",
    "        read sequence\n",
    "    :return softclipped_read_sequence: string\n",
    "    \"\"\"\n",
    "    if right == 0:\n",
    "        return read_seq[left:]\n",
    "    return read_seq[left:-right]\n",
    "\n",
    "\n",
    "def ct_mismatches(read_seq, md, sense):\n",
    "    \"\"\"\n",
    "    Given a read sequence, MD tag, and 'sense' (look for CT if sense,\n",
    "    look for GA if antisense), return the number of CT/GA mismatches\n",
    "    seen in the read.\n",
    "    :param read_seq: string\n",
    "    :param md: string\n",
    "    :param sense: boolean\n",
    "    :return nonAG: int\n",
    "    \"\"\"\n",
    "    mismatches_regex = ur\"(\\d+)([ATCG])\"\n",
    "    mismatches = re.findall(mismatches_regex,md)\n",
    "    ct_mm_counts = 0\n",
    "    if mismatches:\n",
    "        read_pos = 0\n",
    "        for mismatch in mismatches:\n",
    "            ref_allele = mismatch[1]\n",
    "            read_pos += int(mismatch[0])\n",
    "\n",
    "            read_allele = read_seq[read_pos]\n",
    "            if((ref_allele == 'C' and read_allele == 'T' and sense == True) or\n",
    "                   (ref_allele == 'G' and read_allele == 'A' and sense == False)):\n",
    "                ct_mm_counts += 1\n",
    "            read_pos += 1\n",
    "\n",
    "    return ct_mm_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 0 reads\n",
      "Parsed 10000000 reads\n",
      "Parsed 20000000 reads\n",
      "Parsed 30000000 reads\n",
      "Parsed 40000000 reads\n",
      "Parsed 50000000 reads\n",
      "Parsed 60000000 reads\n",
      "Parsed 70000000 reads\n",
      "Parsed 80000000 reads\n",
      "Parsed 90000000 reads\n",
      "Parsed 100000000 reads\n",
      "Parsed 110000000 reads\n",
      "Parsed 120000000 reads\n",
      "Parsed 130000000 reads\n",
      "Parsed 140000000 reads\n",
      "Parsed 150000000 reads\n",
      "Parsed 160000000 reads\n",
      "Parsed 170000000 reads\n",
      "Parsed 180000000 reads\n",
      "Parsed 190000000 reads\n",
      "Parsed 200000000 reads\n",
      "Parsed 210000000 reads\n",
      "Parsed 220000000 reads\n",
      "Parsed 230000000 reads\n",
      "Parsed 240000000 reads\n",
      "Parsed 250000000 reads\n",
      "Parsed 260000000 reads\n",
      "Parsed 270000000 reads\n",
      "Parsed 280000000 reads\n",
      "Parsed 290000000 reads\n",
      "Parsed 300000000 reads\n",
      "Parsed 310000000 reads\n",
      "Parsed 320000000 reads\n",
      "Parsed 330000000 reads\n",
      "Parsed 340000000 reads\n",
      "Parsed 350000000 reads\n",
      "Parsed 360000000 reads\n",
      "Parsed 370000000 reads\n",
      "Parsed 380000000 reads\n",
      "Parsed 390000000 reads\n",
      "Parsed 400000000 reads\n",
      "Parsed 410000000 reads\n",
      "Parsed 420000000 reads\n",
      "Parsed 430000000 reads\n",
      "Parsed 440000000 reads\n",
      "Parsed 450000000 reads\n",
      "Parsed 460000000 reads\n",
      "Parsed 470000000 reads\n",
      "Parsed 480000000 reads\n",
      "Parsed 490000000 reads\n",
      "argument of type 'NoneType' is not iterable\n"
     ]
    }
   ],
   "source": [
    "flags = defaultdict(int) # keep track of filtered reads\n",
    "reverse_stranded = False # 10X is stranded\n",
    "ct_threshold = 0 # remove any read with more than 0 C>T\n",
    "counter = 0 # track read progress\n",
    "\n",
    "# input_bam = os.path.join(input_dir, 'RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD.bam')\n",
    "input_bam = os.path.join(input_dir, 'HEK-NPC-APOBEC-STAMP_possorted_genome_bam_MD.bam')\n",
    "output_bam = os.path.join(output_dir, 'HEK-NPC-APOBEC-STAMP_possorted_genome_bam_MD.noCT.bam')\n",
    "i = pysam.AlignmentFile(input_bam)\n",
    "o = pysam.AlignmentFile(output_bam, \"wb\", template=i)\n",
    "for read in i:\n",
    "    try:\n",
    "        flag = 1  # start out as a 'good' read\n",
    "        cigar = read.cigarstring\n",
    "        if 'X' in cigar or '=' in cigar:\n",
    "            warn_x = True\n",
    "        read_seq = read.query_sequence\n",
    "        read_name = read.query_name\n",
    "        \"\"\"\n",
    "        Throw out unmapped reads\n",
    "        \"\"\"\n",
    "        if read.is_unmapped:\n",
    "            continue  # must be here, otherwise reads won't have CIGAR\n",
    "        try:\n",
    "            md = read.get_tag('MD')\n",
    "        except KeyError:\n",
    "            warn_mm = True\n",
    "            md = ''\n",
    "        \"\"\"\n",
    "        Remove\n",
    "        \"\"\"\n",
    "        \"\"\"\n",
    "        Takes care of soft clipped bases\n",
    "        (remove bases from the read_seq which are soft clipped\n",
    "        to not interfere with mis-alignments downstream)\n",
    "        \"\"\"\n",
    "        left_softclip, right_softclip = get_softclip(cigar)\n",
    "        read_seq = remove_softclipped_reads(\n",
    "            left_softclip, right_softclip, read_seq\n",
    "        )\n",
    "        \n",
    "        \"\"\"\n",
    "        # Manually setting reversed reads to 'sense' strand per truseq\n",
    "        library protocols (Default is truseq reverse stranded)\n",
    "        \"\"\"\n",
    "        if reverse_stranded:\n",
    "            sense = True if read.is_reverse == True else False\n",
    "        else:\n",
    "            sense = True if read.is_reverse == False else False\n",
    "            \n",
    "        \"\"\"\n",
    "        # 5c) If there exists indels, remove them.\n",
    "        \"\"\"\n",
    "        if 'I' in cigar or 'D' in cigar:\n",
    "            flag = 4\n",
    "\n",
    "        elif read.is_secondary:\n",
    "            flag = 5\n",
    "        \n",
    "        elif ct_mismatches(read_seq, md, sense) > ct_threshold:\n",
    "            flag = 7\n",
    "            \n",
    "        flags[flag] += 1\n",
    "        if flag == 1:\n",
    "            o.write(read)\n",
    "        \n",
    "        if counter % 10000000 == 0:\n",
    "            print(\"Parsed {} reads\".format(counter))\n",
    "        counter += 1\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        break\n",
    "o.close()\n",
    "i.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int, {1: 415427692, 4: 6504555, 5: 60940143, 7: 8141740})"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flags"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metrics:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "RBFOX2-TIA1-STAMP_round2E_read1_feature_FB_possorted_genome_bam_MD.bam\n",
    "passed:    788470247\n",
    "indel:     13174402\n",
    "secondary: 126303563\n",
    "edited:    22091603\n",
    "    \n",
    "APOBEC-STAMP_possorted_genome_bam_MD.bam\n",
    "passed:    532823762\n",
    "indel:     10273453\n",
    "secondary: 79750960\n",
    "edited:    10302036\n",
    "\n",
    "RPS2-STAMP_possorted_genome_bam_MD.bam\n",
    "passed:    541507620\n",
    "indel:     10248844\n",
    "secondary: 78460520\n",
    "edited:    19697395\n",
    "\n",
    "HEK-NPC-APOBEC-STAMP_possorted_genome_bam_MD.bam\n",
    "passed:    415427692\n",
    "indel:     6504555\n",
    "secondary: 60940143\n",
    "edited:    8141740"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python2-essential",
   "language": "python",
   "name": "python2-essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
