{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counts the number of C's in a gene's exon or gene location. \n",
    "- Functionally equivalent to the SUPPLEMENT_get_number_of_c.ipynb in 01_bulk except that we need to parse a few annotations from 10X reference data. \n",
    "- We should compare the two just so we know the annotations are more or less the same. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gffutils\n",
    "import pysam\n",
    "import pybedtools\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "from Bio import SeqIO\n",
    "pd.set_option('display.max_columns', 150)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/'\n",
    "output_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/inputs/'\n",
    "genome_fa = '/projects/ps-yeolab4/NCRCRG/refs/refdata-cellranger-hg19-3.0.0/fasta/genome.fa'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grep for exons and genes from the 10X annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = '/projects/ps-yeolab4/NCRCRG/refs/refdata-cellranger-hg19-3.0.0/genes/genes.gtf'\n",
    "exons_gtf = os.path.join(output_dir, 'genes.exons.gtf')\n",
    "genes_gtf = os.path.join(output_dir, 'genes.genes.gtf')\n",
    "\n",
    "! grep exon $genes > $exons_gtf\n",
    "! grep -P '\\tgene\\t' $genes > $genes_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now convert the GTF lines into BED6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab4/software/yeolabconda3/envs/python3essential-0.0.1/lib/python3.6/site-packages/pandas/io/parsers.py:702: UserWarning: Duplicate names specified. This will raise an error in the future.\n",
      "  return _read(filepath_or_buffer, kwds)\n",
      "/projects/ps-yeolab4/software/yeolabconda3/envs/python3essential-0.0.1/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "for fn in [exons_gtf, genes_gtf]:\n",
    "    gtf = pd.read_csv(fn, names=['chrom','src','region','start','end','.','strand','.','attr'], sep='\\t')\n",
    "    gtf['geneid'] = gtf['attr'].str.extract(\"gene_id \\\"([\\w\\d\\.]+)\\\"\")\n",
    "    gtf['score'] = 0\n",
    "    gtf = gtf[['chrom','start','end','geneid','score','strand']]\n",
    "    gtf.to_csv(fn + \".bed\", sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_file = exons_gtf + \".bed\"\n",
    "genes_file = genes_gtf + \".bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_positions_as_df(fn):\n",
    "    \"\"\"\n",
    "    Reads in a BED file and returns as a DataFrame.\n",
    "    \"\"\"\n",
    "    return pd.read_csv(fn, sep='\\t', index_col=3, names=['chrom','start','end','score','strand'])\n",
    "\n",
    "\n",
    "def get_merged_positions_as_bedtool(geneid, df):\n",
    "    \"\"\"\n",
    "    Takes a dataframe (df) and gets regions pertaining only to (geneid). \n",
    "    Does a strand-specific merge and returns non-overlapping regions as a bedtool.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        df = df.loc[geneid].reset_index()[['chrom','start','end','index','score','strand']]\n",
    "        return pybedtools.BedTool.from_dataframe(df).sort().merge(\n",
    "            s=True, c=\"4,5,6\", o=\"distinct,sum,distinct\"\n",
    "        )\n",
    "    except KeyError:\n",
    "        return pybedtools.BedTool.from_dataframe(pd.DataFrame(df.loc[geneid]).T.reset_index()[['chrom','start','end','index','score','strand']]).sort().merge(\n",
    "            s=True, c=\"4,5,6\", o=\"distinct,sum,distinct\"\n",
    "        )\n",
    "    \n",
    "\n",
    "def get_total_editable_sites(bedtool, genome_fa, debug=False):\n",
    "    \"\"\"\n",
    "    Gets total number of total C's in each gene, returns dataframe.\n",
    "    - negative strand returns reverse complement so always look for C\n",
    "    \"\"\"\n",
    "    gene_sequence = bedtool.sequence(fi=genome_fa, s=True, name=True)\n",
    "    total_editable = 0\n",
    "    with open(gene_sequence.seqfn) as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            if debug:\n",
    "                print(\"Sequence: [{}]\".format(record.seq.upper()))\n",
    "            total_editable += record.seq.upper().count('C')\n",
    "            if debug:\n",
    "                print(\"Number Cs found: {}\".format(total_editable))\n",
    "    return total_editable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/ps-yeolab4/software/yeolabconda3/envs/python3essential-0.0.1/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3249: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    }
   ],
   "source": [
    "exons = get_positions_as_df(exons_file)\n",
    "genes = get_positions_as_df(genes_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>27522</td>\n",
       "      <td>27637</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>27522</td>\n",
       "      <td>27614</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>27612</td>\n",
       "      <td>27614</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>26348</td>\n",
       "      <td>26820</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>26416</td>\n",
       "      <td>26820</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215611</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>26413</td>\n",
       "      <td>26415</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      chrom  start    end  score strand\n",
       "ENSG00000215611  GL000201.1  27522  27637      0      -\n",
       "ENSG00000215611  GL000201.1  27522  27614      0      -\n",
       "ENSG00000215611  GL000201.1  27612  27614      0      -\n",
       "ENSG00000215611  GL000201.1  26348  26820      0      -\n",
       "ENSG00000215611  GL000201.1  26416  26820      0      -\n",
       "ENSG00000215611  GL000201.1  26413  26415      0      -"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.loc['ENSG00000215611']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>26348</td>\n",
       "      <td>26820</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000201.1</td>\n",
       "      <td>27522</td>\n",
       "      <td>27637</td>\n",
       "      <td>ENSG00000215611</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  start    end             name  score strand\n",
       "0  GL000201.1  26348  26820  ENSG00000215611      0      -\n",
       "1  GL000201.1  27522  27637  ENSG00000215611      0      -"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_merged_positions_as_bedtool('ENSG00000215611', exons).to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5593def6a26643bc87865f0410da00fa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=32738), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d54d81d2e1d24afcb538e2736f52d33b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=0, max=32738), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Command was:\n",
      "\n",
      "\tbedtools getfasta -s -name -fo /scratch/bay001/21959923.tscc-mgr7.local/pybedtools.7t0ux2i8.tmp -fi /projects/ps-yeolab4/NCRCRG/refs/refdata-cellranger-hg19-3.0.0/fasta/genome.fa -bed /scratch/bay001/21959923.tscc-mgr7.local/pybedtools.4jgetb2r.tmp\n",
      "\n",
      "Error message was:\n",
      "Feature (2:34928447-34928447) has length = 0, Skipping.\n",
      " ENSG00000272027\n"
     ]
    }
   ],
   "source": [
    "editable = defaultdict(dict)\n",
    "\n",
    "for label, region in zip(['genes','exons'], [genes, exons]):\n",
    "    progress = tnrange(len(set(region.index)))\n",
    "    for gene in set(region.index):\n",
    "        try:\n",
    "            bedtool = get_merged_positions_as_bedtool(gene, df=region)\n",
    "            if type(bedtool) == pybedtools.BedTool:\n",
    "                editable[gene][label] = get_total_editable_sites(\n",
    "                    bedtool=bedtool,\n",
    "                    genome_fa=genome_fa,\n",
    "                )\n",
    "            pybedtools.cleanup()\n",
    "        except Exception as e:\n",
    "            print(e, gene)\n",
    "\n",
    "        progress.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(editable).T.to_csv(os.path.join(input_dir, 'genes.editable_C.tsv'), sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exons</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000204832</th>\n",
       "      <td>409.0</td>\n",
       "      <td>5864.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000130812</th>\n",
       "      <td>648.0</td>\n",
       "      <td>2710.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000247765</th>\n",
       "      <td>459.0</td>\n",
       "      <td>4940.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000073969</th>\n",
       "      <td>1014.0</td>\n",
       "      <td>31185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000258947</th>\n",
       "      <td>1853.0</td>\n",
       "      <td>5123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000164509</th>\n",
       "      <td>991.0</td>\n",
       "      <td>15172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000261267</th>\n",
       "      <td>103.0</td>\n",
       "      <td>5653.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237473</th>\n",
       "      <td>112.0</td>\n",
       "      <td>33519.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000168496</th>\n",
       "      <td>583.0</td>\n",
       "      <td>1015.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229563</th>\n",
       "      <td>1036.0</td>\n",
       "      <td>23194.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000233851</th>\n",
       "      <td>129.0</td>\n",
       "      <td>2910.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000147316</th>\n",
       "      <td>1013.0</td>\n",
       "      <td>49459.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272172</th>\n",
       "      <td>86.0</td>\n",
       "      <td>86.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000258888</th>\n",
       "      <td>143.0</td>\n",
       "      <td>764.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000120729</th>\n",
       "      <td>610.0</td>\n",
       "      <td>3832.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000196166</th>\n",
       "      <td>508.0</td>\n",
       "      <td>3985.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000250900</th>\n",
       "      <td>438.0</td>\n",
       "      <td>1983.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000153029</th>\n",
       "      <td>2107.0</td>\n",
       "      <td>6422.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000177084</th>\n",
       "      <td>4331.0</td>\n",
       "      <td>15198.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000240257</th>\n",
       "      <td>110.0</td>\n",
       "      <td>1481.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000116685</th>\n",
       "      <td>987.0</td>\n",
       "      <td>1834.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272076</th>\n",
       "      <td>85.0</td>\n",
       "      <td>85.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000006555</th>\n",
       "      <td>1810.0</td>\n",
       "      <td>5838.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000197670</th>\n",
       "      <td>494.0</td>\n",
       "      <td>5276.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198169</th>\n",
       "      <td>1275.0</td>\n",
       "      <td>8276.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000104818</th>\n",
       "      <td>302.0</td>\n",
       "      <td>414.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000230430</th>\n",
       "      <td>448.0</td>\n",
       "      <td>448.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000136937</th>\n",
       "      <td>1131.0</td>\n",
       "      <td>7318.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000161048</th>\n",
       "      <td>1439.0</td>\n",
       "      <td>9719.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000231791</th>\n",
       "      <td>70.0</td>\n",
       "      <td>755.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158220</th>\n",
       "      <td>2218.0</td>\n",
       "      <td>11092.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000261298</th>\n",
       "      <td>310.0</td>\n",
       "      <td>310.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000267338</th>\n",
       "      <td>76.0</td>\n",
       "      <td>135.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000185973</th>\n",
       "      <td>836.0</td>\n",
       "      <td>33456.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000144802</th>\n",
       "      <td>1315.0</td>\n",
       "      <td>6271.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000196407</th>\n",
       "      <td>305.0</td>\n",
       "      <td>1755.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000136014</th>\n",
       "      <td>1029.0</td>\n",
       "      <td>6948.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000269220</th>\n",
       "      <td>610.0</td>\n",
       "      <td>610.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000163507</th>\n",
       "      <td>779.0</td>\n",
       "      <td>6558.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000129465</th>\n",
       "      <td>982.0</td>\n",
       "      <td>1217.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000226694</th>\n",
       "      <td>89.0</td>\n",
       "      <td>1910.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000150510</th>\n",
       "      <td>1436.0</td>\n",
       "      <td>13883.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000254286</th>\n",
       "      <td>278.0</td>\n",
       "      <td>31987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000257920</th>\n",
       "      <td>142.0</td>\n",
       "      <td>769.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000116752</th>\n",
       "      <td>221.0</td>\n",
       "      <td>2666.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000215343</th>\n",
       "      <td>664.0</td>\n",
       "      <td>1971.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000224459</th>\n",
       "      <td>189.0</td>\n",
       "      <td>2861.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000125166</th>\n",
       "      <td>901.0</td>\n",
       "      <td>6045.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000166436</th>\n",
       "      <td>2855.0</td>\n",
       "      <td>13237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000130876</th>\n",
       "      <td>968.0</td>\n",
       "      <td>4776.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000171840</th>\n",
       "      <td>634.0</td>\n",
       "      <td>24510.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272593</th>\n",
       "      <td>156.0</td>\n",
       "      <td>156.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000105261</th>\n",
       "      <td>300.0</td>\n",
       "      <td>685.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000226894</th>\n",
       "      <td>670.0</td>\n",
       "      <td>5469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228361</th>\n",
       "      <td>104.0</td>\n",
       "      <td>2047.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000124343</th>\n",
       "      <td>658.0</td>\n",
       "      <td>13759.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000165219</th>\n",
       "      <td>2685.0</td>\n",
       "      <td>21337.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000269640</th>\n",
       "      <td>615.0</td>\n",
       "      <td>3308.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000140107</th>\n",
       "      <td>596.0</td>\n",
       "      <td>2056.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000061337</th>\n",
       "      <td>1776.0</td>\n",
       "      <td>14393.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  exons    genes\n",
       "ENSG00000204832   409.0   5864.0\n",
       "ENSG00000130812   648.0   2710.0\n",
       "ENSG00000247765   459.0   4940.0\n",
       "ENSG00000073969  1014.0  31185.0\n",
       "ENSG00000258947  1853.0   5123.0\n",
       "ENSG00000164509   991.0  15172.0\n",
       "ENSG00000261267   103.0   5653.0\n",
       "ENSG00000237473   112.0  33519.0\n",
       "ENSG00000168496   583.0   1015.0\n",
       "ENSG00000229563  1036.0  23194.0\n",
       "ENSG00000233851   129.0   2910.0\n",
       "ENSG00000147316  1013.0  49459.0\n",
       "ENSG00000272172    86.0     86.0\n",
       "ENSG00000258888   143.0    764.0\n",
       "ENSG00000120729   610.0   3832.0\n",
       "ENSG00000196166   508.0   3985.0\n",
       "ENSG00000250900   438.0   1983.0\n",
       "ENSG00000153029  2107.0   6422.0\n",
       "ENSG00000177084  4331.0  15198.0\n",
       "ENSG00000240257   110.0   1481.0\n",
       "ENSG00000116685   987.0   1834.0\n",
       "ENSG00000272076    85.0     85.0\n",
       "ENSG00000006555  1810.0   5838.0\n",
       "ENSG00000197670   494.0   5276.0\n",
       "ENSG00000198169  1275.0   8276.0\n",
       "ENSG00000104818   302.0    414.0\n",
       "ENSG00000230430   448.0    448.0\n",
       "ENSG00000136937  1131.0   7318.0\n",
       "ENSG00000161048  1439.0   9719.0\n",
       "ENSG00000231791    70.0    755.0\n",
       "...                 ...      ...\n",
       "ENSG00000158220  2218.0  11092.0\n",
       "ENSG00000261298   310.0    310.0\n",
       "ENSG00000267338    76.0    135.0\n",
       "ENSG00000185973   836.0  33456.0\n",
       "ENSG00000144802  1315.0   6271.0\n",
       "ENSG00000196407   305.0   1755.0\n",
       "ENSG00000136014  1029.0   6948.0\n",
       "ENSG00000269220   610.0    610.0\n",
       "ENSG00000163507   779.0   6558.0\n",
       "ENSG00000129465   982.0   1217.0\n",
       "ENSG00000226694    89.0   1910.0\n",
       "ENSG00000150510  1436.0  13883.0\n",
       "ENSG00000254286   278.0  31987.0\n",
       "ENSG00000257920   142.0    769.0\n",
       "ENSG00000116752   221.0   2666.0\n",
       "ENSG00000215343   664.0   1971.0\n",
       "ENSG00000224459   189.0   2861.0\n",
       "ENSG00000125166   901.0   6045.0\n",
       "ENSG00000166436  2855.0  13237.0\n",
       "ENSG00000130876   968.0   4776.0\n",
       "ENSG00000171840   634.0  24510.0\n",
       "ENSG00000272593   156.0    156.0\n",
       "ENSG00000105261   300.0    685.0\n",
       "ENSG00000226894   670.0   5469.0\n",
       "ENSG00000228361   104.0   2047.0\n",
       "ENSG00000124343   658.0  13759.0\n",
       "ENSG00000165219  2685.0  21337.0\n",
       "ENSG00000269640   615.0   3308.0\n",
       "ENSG00000140107   596.0   2056.0\n",
       "ENSG00000061337  1776.0  14393.0\n",
       "\n",
       "[32738 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(editable).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3-essential-tscc",
   "language": "python",
   "name": "python3essentialtscc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
