{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook contains identical code to SAILOR's read filter. \n",
    "- Doing this because there are some reads in our filtered set that don't seem to also be in the 'all' set??? Let's figure out why."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam \n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/sailor_outputs_groups/original_outputs/RPS2-STAMP_possorted_genome_bam_MD/results'\n",
    "output_dir = '/oasis/tscc/scratch/bay001/sailor/'\n",
    "\n",
    "input_bam = os.path.join(input_dir, 'RPS2-STAMP_possorted_genome_bam_MD.rev.sorted.rmdup.bam')\n",
    "output_bam = os.path.join(output_dir, os.path.basename(input_bam) + \".filtered.bam\")\n",
    "\n",
    "assert os.path.exists(input_bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_softclip(cigar):\n",
    "    \"\"\"\n",
    "    Returns the number of bases to be softclipped on either left or right\n",
    "    side. Or both. If not softclipped, return 0\n",
    "\n",
    "    :param cigar: string\n",
    "        BAM/SAM CIGAR string\n",
    "    :return left: int\n",
    "        number of softclipped reads at the beginning\n",
    "    :return right: int\n",
    "        number of softclipped reads at the end\n",
    "    \"\"\"\n",
    "    softclip_regex = ur\"(\\d+)S\"\n",
    "    softclip = re.findall(softclip_regex,cigar)\n",
    "\n",
    "    softclip_right_regex = ur\"[\\w\\d]{1}(\\d+)S\" # if the softclip comes from the right side\n",
    "    softclip_right = re.findall(softclip_right_regex,cigar)\n",
    "\n",
    "    left = 0\n",
    "    right = 0\n",
    "\n",
    "    if softclip:\n",
    "        if len(softclip) == 2: # softclipped on both sides\n",
    "            left = int(softclip[0])\n",
    "            right = int(softclip[1])\n",
    "        elif len(softclip_right) == 1: # softclipped only on the RIGHT side\n",
    "            right = int(softclip[0])\n",
    "        else: # softclipped only on the LEFT side\n",
    "            left = int(softclip[0])\n",
    "    return left, right\n",
    "\n",
    "\n",
    "def remove_softclipped_reads(left, right, read_seq):\n",
    "    \"\"\"\n",
    "    Returns the read after removing softclipped bases.\n",
    "    :param left: int\n",
    "        left softclip\n",
    "    :param right: int\n",
    "        right softclip\n",
    "    :param read_seq: string\n",
    "        read sequence\n",
    "    :return softclipped_read_sequence: string\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    if right == 0:\n",
    "        return read_seq[left:]\n",
    "    return read_seq[left:-right]\n",
    "\n",
    "\n",
    "def get_junction_overhangs(cigar, min_jct_overhang):\n",
    "    \"\"\"\n",
    "    Returns the MIN number of reads left/right of ANY junction\n",
    "    as indicated by the N in a cigar string.\n",
    "    Return -1, -1 for reads that don't span junctions.\n",
    "\n",
    "    :param cigar: string\n",
    "    :param min_jct_overhang: int\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if cigar.count('N') == 1:\n",
    "        return get_single_junction_overhang(cigar)\n",
    "\n",
    "    min_left, min_right = get_single_junction_overhang(cigar)\n",
    "    for i in [m.start() for m in re.finditer(r\"N\",cigar)]:\n",
    "        sub_cigar = cigar[i+1:]\n",
    "        left, right = get_single_junction_overhang(sub_cigar)\n",
    "        if left != -1 and right != -1:\n",
    "            min_left = min(min_left, left)\n",
    "            min_right = min(min_right, right)\n",
    "    return min_left, min_right\n",
    "\n",
    "\n",
    "def get_single_junction_overhang(cigar):\n",
    "    \"\"\"\n",
    "    Returns the number of reads left/right of a junction as indicated\n",
    "    by the LEFTMOST N in a cigar string. Return -1, -1 for reads that don't span\n",
    "    junctions.\n",
    "\n",
    "    :param cigar: string\n",
    "    :return left: int\n",
    "    :return right: int\n",
    "    \"\"\"\n",
    "    cigar_overhang_regex = ur\"(\\d+)M[\\d]+N(\\d+)M\"\n",
    "\n",
    "    overhangs = re.findall(cigar_overhang_regex, cigar)\n",
    "    if overhangs:\n",
    "        return int(overhangs[0][0]), int(overhangs[0][1])\n",
    "    else:\n",
    "        return -1, -1\n",
    "\n",
    "\n",
    "def is_mismatch_before_n_flank_of_read(md, n):\n",
    "    \"\"\"\n",
    "    Returns True if there is a mismatch before the first n nucleotides\n",
    "    of a read, or if there is a mismatch before the last n nucleotides\n",
    "    of a read.\n",
    "\n",
    "    :param md: string\n",
    "    :param n: int\n",
    "    :return is_mismatch: boolean\n",
    "    \"\"\"\n",
    "    is_mismatch = False\n",
    "    flank_mm_regex = ur\"^(\\d+).*[ACGT](\\d+)$\"\n",
    "    flank_mm = re.findall(flank_mm_regex,md)\n",
    "    if flank_mm:\n",
    "        flank_mm = flank_mm[0]\n",
    "        if flank_mm[1]:\n",
    "            if int(flank_mm[1]) < n:\n",
    "                is_mismatch = True\n",
    "        if flank_mm[0]:\n",
    "            if int(flank_mm[0]) < n:\n",
    "                is_mismatch = True\n",
    "    return is_mismatch\n",
    "\n",
    "\n",
    "def non_ag_mismatches(read_seq, md, sense):\n",
    "    \"\"\"\n",
    "    Given a read sequence, MD tag, and 'sense' (look for AG if sense,\n",
    "    look for TC if antisense), return the number of non-AG/TC mismatches\n",
    "    seen in the read.\n",
    "\n",
    "    :param read_seq: string\n",
    "    :param md: string\n",
    "    :param sense: boolean\n",
    "    :return nonAG: int\n",
    "    \"\"\"\n",
    "    mismatches_regex = ur\"(\\d+)([ATCG])\"\n",
    "    mismatches = re.findall(mismatches_regex,md)\n",
    "    non_ag_mm_counts = 0\n",
    "    if mismatches:\n",
    "        read_pos = 0\n",
    "        for mismatch in mismatches:\n",
    "            ref_allele = mismatch[1]\n",
    "            read_pos += int(mismatch[0])\n",
    "\n",
    "            read_allele = read_seq[read_pos]\n",
    "            if(not((ref_allele == 'A' and read_allele == 'G' and sense == True) or\n",
    "                   (ref_allele == 'T' and read_allele == 'C' and sense == False))):\n",
    "                non_ag_mm_counts += 1\n",
    "            read_pos += 1\n",
    "\n",
    "    return non_ag_mm_counts\n",
    "\n",
    "\n",
    "def non_ct_mismatches(read_seq, md, sense):\n",
    "    \"\"\"\n",
    "    Given a read sequence, MD tag, and 'sense' (look for CT if sense,\n",
    "    look for GA if antisense), return the number of non-CT/GA mismatches\n",
    "    seen in the read.\n",
    "\n",
    "    :param read_seq: string\n",
    "    :param md: string\n",
    "    :param sense: boolean\n",
    "    :return nonAG: int\n",
    "    \"\"\"\n",
    "    mismatches_regex = ur\"(\\d+)([ATCG])\"\n",
    "    mismatches = re.findall(mismatches_regex,md)\n",
    "    non_ct_mm_counts = 0\n",
    "    if mismatches:\n",
    "        read_pos = 0\n",
    "        for mismatch in mismatches:\n",
    "            ref_allele = mismatch[1]\n",
    "            read_pos += int(mismatch[0])\n",
    "\n",
    "            read_allele = read_seq[read_pos]\n",
    "            if(not((ref_allele == 'C' and read_allele == 'T' and sense == True) or\n",
    "                   (ref_allele == 'G' and read_allele == 'A' and sense == False))):\n",
    "                non_ct_mm_counts += 1\n",
    "            read_pos += 1\n",
    "\n",
    "    return non_ct_mm_counts\n",
    "\n",
    "\n",
    "def non_gt_mismatches(read_seq, md, sense):\n",
    "    \"\"\"\n",
    "    Given a read sequence, MD tag, and 'sense' (look for GT if sense,\n",
    "    look for CA if antisense), return the number of non-GT/CA mismatches\n",
    "    seen in the read.\n",
    "\n",
    "    :param read_seq: string\n",
    "    :param md: string\n",
    "    :param sense: boolean\n",
    "    :return nonAG: int\n",
    "    \"\"\"\n",
    "    mismatches_regex = ur\"(\\d+)([ATCG])\"\n",
    "    mismatches = re.findall(mismatches_regex,md)\n",
    "    non_gt_mm_counts = 0\n",
    "    if mismatches:\n",
    "        read_pos = 0\n",
    "        for mismatch in mismatches:\n",
    "            ref_allele = mismatch[1]\n",
    "            read_pos += int(mismatch[0])\n",
    "\n",
    "            read_allele = read_seq[read_pos]\n",
    "            if(not((ref_allele == 'G' and read_allele == 'T' and sense == True) or\n",
    "                   (ref_allele == 'C' and read_allele == 'A' and sense == False))):\n",
    "                non_gt_mm_counts += 1\n",
    "            read_pos += 1\n",
    "\n",
    "    return non_gt_mm_counts\n",
    "\n",
    "\n",
    "def filter_reads(\n",
    "        input_bam, output_bam, min_overhang, min_underhang,\n",
    "        non_ag_mm_threshold, reverse_stranded=True, ct=False, gt=False\n",
    "):\n",
    "    \"\"\"\n",
    "    # Step 3: filter reads\n",
    "\n",
    "    Per Boyko/Mike Washburn's editing paper:\n",
    "\n",
    "    b) it had a junction overhang < 10nt according to its SAMtools CIGAR string\n",
    "    c) it had > 1 non-A2G or non-C2T mismatch or any short indel, per its MD tag\n",
    "    d) it had a mismatch less than 25nt away from either end of the read\n",
    "    (this was changed to 5nt in the relaxed version used for quantification)\n",
    "\n",
    "    * Removed 5d per discussions with the Hundley lab.\n",
    "\n",
    "    :param input_bam: basestring\n",
    "    :param output_bam: basestring\n",
    "    :param min_overhang: int\n",
    "        minimum number required to span a junction (filter b)\n",
    "    :param min_underhang: int\n",
    "        minimum number required at the end of the read (filter c)\n",
    "    :param non_ag_mm_threshold: int\n",
    "        any more non A-G mismatches than this will filter the read.\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    print(\"Filtering reads on: {}\".format(input_bam))\n",
    "    i = pysam.AlignmentFile(input_bam)\n",
    "    o = pysam.AlignmentFile(output_bam, \"wb\", template=i)\n",
    "    flags = defaultdict(list) # number of flags in the bam file\n",
    "    warn_mm = False\n",
    "    warn_junction = False\n",
    "    warn_x = False\n",
    "\n",
    "    counter = 0\n",
    "    for read in i:\n",
    "        if counter % 1000000 == 0:\n",
    "            print(\"Parsed {} reads\".format(counter))\n",
    "        try:\n",
    "            flag = 1  # start out as a 'good' read\n",
    "            cigar = read.cigarstring\n",
    "            if 'X' in cigar or '=' in cigar:\n",
    "                warn_x = True\n",
    "            read_seq = read.query_sequence\n",
    "            read_name = read.query_name\n",
    "            \"\"\"\n",
    "            Throw out unmapped reads\n",
    "            \"\"\"\n",
    "            if read.is_unmapped:\n",
    "                flags['unmapped'].append(read.name)\n",
    "                continue  # must be here, otherwise reads won't have CIGAR\n",
    "            try:\n",
    "                mm = read.get_tag('MD')\n",
    "            except KeyError:\n",
    "                warn_mm = True\n",
    "                mm = ''\n",
    "            \"\"\"\n",
    "            Remove\n",
    "            \"\"\"\n",
    "            \"\"\"\n",
    "            Takes care of soft clipped bases\n",
    "            (remove bases from the read_seq which are soft clipped\n",
    "            to not interfere with mis-alignments downstream)\n",
    "            \"\"\"\n",
    "            left_softclip, right_softclip = get_softclip(cigar)\n",
    "            read_seq = remove_softclipped_reads(\n",
    "                left_softclip, right_softclip, read_seq\n",
    "            )\n",
    "\n",
    "            \"\"\"\n",
    "            # 5b) Check for small junction overhangs.\n",
    "            If match junction over hang is small, (if mismatches occur close to\n",
    "            junctions), remove read\n",
    "            \"\"\"\n",
    "\n",
    "            left_overhang, right_overhang = get_junction_overhangs(cigar, min_overhang)\n",
    "            if left_overhang != -1 and right_overhang != -1:\n",
    "                if left_overhang < min_overhang or right_overhang < min_overhang:\n",
    "                    flags['small_overhang'].append(read_name)\n",
    "                    flag = 3\n",
    "            elif left_overhang != -1 or right_overhang != -1:\n",
    "                warn_junction = True\n",
    "\n",
    "\n",
    "            \"\"\"\n",
    "            # 5c) If there exists indels, remove them.\n",
    "            \"\"\"\n",
    "            if 'I' in cigar or 'D' in cigar:\n",
    "                flags['indel'].append(read_name)\n",
    "                flag = 4\n",
    "\n",
    "            \"\"\"\n",
    "            # 2) If primary not primary alignment, throw out\n",
    "            \"\"\"\n",
    "    \n",
    "            if read.is_secondary:\n",
    "                flags['not_primary'].append(read_name)\n",
    "                flag = 5\n",
    "\n",
    "            \"\"\"\n",
    "            # MD:Z tag-based filters. If there is a mismatch on either end of the read, throw out.\n",
    "            \"\"\"\n",
    "\n",
    "            if is_mismatch_before_n_flank_of_read(mm, min_underhang):\n",
    "                flags['small_underhang'].append(read_name)\n",
    "                flag = 6\n",
    "\n",
    "            \"\"\"\n",
    "            # Manually setting reversed reads to 'sense' strand per truseq\n",
    "            library protocols (Default is truseq reverse stranded)\n",
    "            \"\"\"\n",
    "            if reverse_stranded:\n",
    "                sense = True if read.is_reverse == True else False\n",
    "            else:\n",
    "                sense = True if read.is_reverse == False else False\n",
    "\n",
    "            \"\"\"\n",
    "            # 5c) Search MDZ for A, T's in reference, if mutations are not \n",
    "            #     A-G (sense) or T-C (antisense), add to non_ag_mm_counts\n",
    "            #     threshold. If non_ag_mm_counts > threshold, toss the whole\n",
    "            #     read. Otherwise, allow up to the maximum allowable non-AG\n",
    "            #     mismatches before tossing. Default: 1 mm\n",
    "            \"\"\"\n",
    "            if ct:\n",
    "                if non_ct_mismatches(read_seq, mm, sense) > non_ag_mm_threshold:\n",
    "                    flags['non_ct_exceeded'].append(read_name)\n",
    "                    flag = 7\n",
    "            elif gt:\n",
    "                if non_gt_mismatches(read_seq, mm, sense) > non_ag_mm_threshold:\n",
    "                    flags['non_gt_exceeded'].append(read_name)\n",
    "                    flag = 7\n",
    "            else:\n",
    "                if non_ag_mismatches(read_seq, mm, sense) > non_ag_mm_threshold:\n",
    "                    flags['non_ag_exceeded'].append(read_name)\n",
    "                    flag = 7\n",
    "\n",
    "            if flag == 1:\n",
    "                o.write(read)\n",
    "                \n",
    "            if read_name == 'A00953:41:HWMCFDSXX:1:1259:28800:4617':\n",
    "                print(\"Flag for {} is {}\".format(\"A00953:41:HWMCFDSXX:1:1259:28800:4617\", flag))\n",
    "        except TypeError as e:\n",
    "            print(\"error! {}\".format(e))\n",
    "            \n",
    "        counter += 1\n",
    "    i.close()\n",
    "    o.close()\n",
    "    # warn if optional MD tag missing from any read\n",
    "    if warn_mm:\n",
    "        print(\"Warning, reads lack optional MD: tag (could not calculate non-ag mismatches)\")\n",
    "    if warn_junction:\n",
    "        print(\"Warning: CIGAR {} has weird junction mark (or the regex is wrong)\")\n",
    "    if warn_x:\n",
    "        print(\"Warning: I don't like X or = in CIGAR (use M instead)\")\n",
    "\n",
    "    return flags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering reads on: /home/bay001/projects/kris_apobec_20200121/permanent_data2/07_scRNA_groups/sailor_outputs_groups/original_outputs/RPS2-STAMP_possorted_genome_bam_MD/results/RPS2-STAMP_possorted_genome_bam_MD.rev.sorted.rmdup.bam\n",
      "Parsed 0 reads\n",
      "Parsed 1000000 reads\n",
      "Parsed 2000000 reads\n",
      "Parsed 3000000 reads\n",
      "Parsed 4000000 reads\n",
      "Parsed 5000000 reads\n",
      "Parsed 6000000 reads\n",
      "Parsed 7000000 reads\n",
      "Parsed 8000000 reads\n",
      "Parsed 9000000 reads\n",
      "Parsed 10000000 reads\n",
      "Parsed 11000000 reads\n",
      "Parsed 12000000 reads\n",
      "Parsed 13000000 reads\n",
      "Parsed 14000000 reads\n",
      "Parsed 15000000 reads\n",
      "Parsed 16000000 reads\n",
      "Parsed 17000000 reads\n",
      "Parsed 18000000 reads\n",
      "Parsed 19000000 reads\n",
      "Parsed 20000000 reads\n",
      "Parsed 21000000 reads\n",
      "Parsed 22000000 reads\n",
      "Parsed 23000000 reads\n",
      "Parsed 24000000 reads\n",
      "Parsed 25000000 reads\n",
      "Parsed 26000000 reads\n",
      "Parsed 27000000 reads\n",
      "Parsed 28000000 reads\n",
      "Parsed 29000000 reads\n",
      "Parsed 30000000 reads\n",
      "Parsed 31000000 reads\n",
      "Parsed 32000000 reads\n",
      "Parsed 33000000 reads\n",
      "Parsed 34000000 reads\n",
      "Parsed 35000000 reads\n",
      "Parsed 36000000 reads\n",
      "Parsed 37000000 reads\n",
      "Parsed 38000000 reads\n",
      "Parsed 39000000 reads\n",
      "Parsed 40000000 reads\n",
      "Parsed 41000000 reads\n",
      "Parsed 42000000 reads\n",
      "Parsed 43000000 reads\n",
      "Parsed 44000000 reads\n",
      "Parsed 45000000 reads\n",
      "Parsed 46000000 reads\n",
      "Parsed 47000000 reads\n",
      "Parsed 48000000 reads\n",
      "Parsed 49000000 reads\n",
      "Parsed 50000000 reads\n",
      "Parsed 51000000 reads\n",
      "Parsed 52000000 reads\n",
      "Parsed 53000000 reads\n",
      "Parsed 54000000 reads\n",
      "Parsed 55000000 reads\n",
      "Parsed 56000000 reads\n",
      "Parsed 57000000 reads\n",
      "Parsed 58000000 reads\n",
      "Parsed 59000000 reads\n",
      "Parsed 60000000 reads\n",
      "Parsed 61000000 reads\n",
      "Parsed 62000000 reads\n",
      "Parsed 63000000 reads\n",
      "Parsed 64000000 reads\n",
      "Parsed 65000000 reads\n",
      "Parsed 66000000 reads\n",
      "Parsed 67000000 reads\n",
      "Parsed 68000000 reads\n",
      "Parsed 69000000 reads\n",
      "Parsed 70000000 reads\n",
      "Parsed 71000000 reads\n",
      "Parsed 72000000 reads\n",
      "Parsed 73000000 reads\n",
      "Parsed 74000000 reads\n",
      "Parsed 75000000 reads\n",
      "Parsed 76000000 reads\n",
      "Parsed 77000000 reads\n",
      "Parsed 78000000 reads\n",
      "Parsed 79000000 reads\n",
      "Parsed 80000000 reads\n",
      "Parsed 81000000 reads\n",
      "Parsed 82000000 reads\n",
      "Parsed 83000000 reads\n",
      "Parsed 84000000 reads\n",
      "Parsed 85000000 reads\n",
      "Parsed 86000000 reads\n",
      "Parsed 87000000 reads\n",
      "Parsed 88000000 reads\n",
      "Parsed 89000000 reads\n",
      "Parsed 90000000 reads\n",
      "Parsed 91000000 reads\n",
      "Parsed 92000000 reads\n",
      "Parsed 93000000 reads\n",
      "Parsed 94000000 reads\n",
      "Parsed 95000000 reads\n",
      "Parsed 96000000 reads\n",
      "Parsed 97000000 reads\n",
      "Parsed 98000000 reads\n",
      "Parsed 99000000 reads\n",
      "Parsed 100000000 reads\n",
      "Parsed 101000000 reads\n",
      "Parsed 102000000 reads\n",
      "Parsed 103000000 reads\n",
      "Parsed 104000000 reads\n",
      "Parsed 105000000 reads\n",
      "Parsed 106000000 reads\n",
      "Parsed 107000000 reads\n",
      "Parsed 108000000 reads\n",
      "Parsed 109000000 reads\n",
      "Parsed 110000000 reads\n",
      "Parsed 111000000 reads\n",
      "Parsed 112000000 reads\n",
      "Parsed 113000000 reads\n",
      "Parsed 114000000 reads\n",
      "Parsed 115000000 reads\n",
      "Parsed 116000000 reads\n",
      "Parsed 117000000 reads\n",
      "Parsed 118000000 reads\n",
      "Parsed 119000000 reads\n",
      "Parsed 120000000 reads\n",
      "Parsed 121000000 reads\n",
      "Parsed 122000000 reads\n",
      "Parsed 123000000 reads\n",
      "Parsed 124000000 reads\n",
      "Parsed 125000000 reads\n",
      "Parsed 126000000 reads\n",
      "Parsed 127000000 reads\n",
      "Parsed 128000000 reads\n",
      "Parsed 129000000 reads\n",
      "Parsed 130000000 reads\n",
      "Parsed 131000000 reads\n",
      "Parsed 132000000 reads\n",
      "Parsed 133000000 reads\n",
      "Parsed 134000000 reads\n",
      "Parsed 135000000 reads\n",
      "Flag for A00953:41:HWMCFDSXX:1:1259:28800:4617 is 6\n",
      "Parsed 136000000 reads\n",
      "Parsed 137000000 reads\n",
      "Parsed 138000000 reads\n",
      "Parsed 139000000 reads\n",
      "Parsed 140000000 reads\n",
      "Parsed 141000000 reads\n",
      "Parsed 142000000 reads\n",
      "Parsed 143000000 reads\n",
      "Parsed 144000000 reads\n",
      "Parsed 145000000 reads\n",
      "Parsed 146000000 reads\n",
      "Parsed 147000000 reads\n",
      "Parsed 148000000 reads\n",
      "Parsed 149000000 reads\n",
      "Parsed 150000000 reads\n",
      "Parsed 151000000 reads\n",
      "Parsed 152000000 reads\n",
      "Parsed 153000000 reads\n",
      "Parsed 154000000 reads\n",
      "Parsed 155000000 reads\n",
      "Parsed 156000000 reads\n",
      "Parsed 157000000 reads\n",
      "Parsed 158000000 reads\n",
      "Parsed 159000000 reads\n",
      "Parsed 160000000 reads\n",
      "Parsed 161000000 reads\n",
      "Parsed 162000000 reads\n",
      "Parsed 163000000 reads\n",
      "Parsed 164000000 reads\n",
      "Parsed 165000000 reads\n",
      "Parsed 166000000 reads\n",
      "Parsed 167000000 reads\n",
      "Parsed 168000000 reads\n",
      "Parsed 169000000 reads\n",
      "Parsed 170000000 reads\n",
      "Parsed 171000000 reads\n",
      "Parsed 172000000 reads\n",
      "Parsed 173000000 reads\n",
      "Parsed 174000000 reads\n",
      "Parsed 175000000 reads\n",
      "Parsed 176000000 reads\n",
      "Parsed 177000000 reads\n",
      "Parsed 178000000 reads\n",
      "Parsed 179000000 reads\n",
      "Parsed 180000000 reads\n",
      "Parsed 181000000 reads\n",
      "Parsed 182000000 reads\n",
      "Parsed 183000000 reads\n",
      "Parsed 184000000 reads\n",
      "Parsed 185000000 reads\n",
      "Parsed 186000000 reads\n",
      "Parsed 187000000 reads\n",
      "Parsed 188000000 reads\n",
      "Parsed 189000000 reads\n",
      "Parsed 190000000 reads\n",
      "Parsed 191000000 reads\n",
      "Parsed 192000000 reads\n",
      "Parsed 193000000 reads\n",
      "Parsed 194000000 reads\n",
      "Parsed 195000000 reads\n",
      "Parsed 196000000 reads\n",
      "Parsed 197000000 reads\n",
      "Parsed 198000000 reads\n",
      "Parsed 199000000 reads\n",
      "Parsed 200000000 reads\n",
      "Parsed 201000000 reads\n",
      "Parsed 202000000 reads\n",
      "Parsed 203000000 reads\n",
      "Parsed 204000000 reads\n",
      "Parsed 205000000 reads\n",
      "Parsed 206000000 reads\n",
      "Parsed 207000000 reads\n",
      "Parsed 208000000 reads\n",
      "Parsed 209000000 reads\n",
      "Parsed 210000000 reads\n",
      "Parsed 211000000 reads\n"
     ]
    }
   ],
   "source": [
    "flags = filter_reads(\n",
    "    input_bam, output_bam, min_overhang=10, min_underhang=5,\n",
    "    non_ag_mm_threshold=1, reverse_stranded=False, ct=True, gt=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['small_overhang', 'non_ct_exceeded', 'small_underhang', 'indel']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flags.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6961507"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flags['small_overhang'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9137940"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flags['non_ct_exceeded'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3798831"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(flags['indel'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python2-essential",
   "language": "python",
   "name": "python2-essential"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
